

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Part 1 - Introduction to Biomedical Networks &#8212; ISMB Tutorial Book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Section1/Part1';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Part 2 - Patient similarity networks" href="Part2.html" />
    <link rel="prev" title="Section 1 - Network Basics" href="../Section1_intro.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="ISMB Tutorial Book - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="ISMB Tutorial Book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    An Introduction to Multi Modal Analysis using Networks
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../GettingStarted.html">Getting Started</a></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../Section1_intro.html">Section 1 - Network Basics</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Part 1 - Introduction to Biomedical Networks</a></li>
















<li class="toctree-l2"><a class="reference internal" href="Part2.html">Part 2 - Patient similarity networks</a></li>

</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section2_intro.html">Section 2: Network Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section2/Part1.html">Part 1: Over-representation and Enrichment Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section2/Part2.html">Part 2: Similarity Network Fusion</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section3_intro.html">Section 3 - SNF + GNNâ€™s</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part1.html"><font color="darkblue"> Part 1 : Message Passing </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part2.html"><font color="darkblue"> Part 2 : Problem Formulation / Patient Similarity Network Generation </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part3.html"><font color="darkblue"> Part 3 : Smoking Network Classification using Graph Convolutional Networks</font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part4.html">Part 4 : Similarity Network Fusion with Graph Convolutional Networks</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ConceptGlossary.html">Concept Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions_definitions.html">Custom Function Definitions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/biomedicalinformaticsgroup/networks_ismb24.git" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/biomedicalinformaticsgroup/networks_ismb24.git/issues/new?title=Issue%20on%20page%20%2FSection1/Part1.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Section1/Part1.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Part 1 - Introduction to Biomedical Networks</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Part 1 - Introduction to Biomedical Networks</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structuring-biomedical-data-using-networks-workshop-network-generation">Structuring Biomedical Data Using Networks  workshop - Network Generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#biomedical-networks">1. Biomedical Networks</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#libraries">1.1 Libraries</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcga-luad-project">1.2 TCGA-LUAD Project</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata-analysis">1.3 Metadata Analysis</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-data-analysis">Expression data analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-filtering-and-preprocessing">2. Data Filtering and Preprocessing</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysing-gene-expression-levels">2.1 Analysing Gene Expression Levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholding-based-on-gene-expression-variance">2.3 Thresholding Based on Gene Expression Variance</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#task-1-variance-based-gene-filtering">Task 1: Variance-Based Gene Filtering</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-calculate-variance">Step 1: Calculate Variance</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-set-a-variance-threshold">Step 2: Set a Variance Threshold</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-filter-genes">Step 3: Filter Genes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-retention-at-different-variance-levels">2.4 Gene Retention at Different Variance Levels</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-symbol-conversion-ensg-to-hgcn">3. Gene Symbol Conversion: ENSG to HGCN</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-data-investigation">Further Data Investigation</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#correlation-matrix-calculation">4. Correlation Matrix Calculation</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-construction-from-correlation-matrices">5. Graph Construction from Correlation Matrices</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#network-cleaning">6. Network Cleaning</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#sparsification-methods">7. Sparsification methods</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#describing-highly-connected-nodes">8. Describing Highly Connected Nodes</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#interactive-visualization">Interactive Visualization</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-add-hover-info-to-plotly-plot-gene-info">TODO: add hover info to plotly plot (gene info)</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-introduce-clustering-kmeans-etc">TODO: introduce clustering kmeans. etc</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#storing-generated-networks">Storing generated networks</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#patient-similarity-network">Patient similarity network</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#dnameth-pn">DNAmeth PN</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-pip-network-using-string">Setting up PIP network using String</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="part-1-introduction-to-biomedical-networks">
<h1>Part 1 - Introduction to Biomedical Networks<a class="headerlink" href="#part-1-introduction-to-biomedical-networks" title="Permalink to this heading">#</a></h1>
<section id="structuring-biomedical-data-using-networks-workshop-network-generation">
<h2>Structuring Biomedical Data Using Networks  workshop - Network Generation<a class="headerlink" href="#structuring-biomedical-data-using-networks-workshop-network-generation" title="Permalink to this heading">#</a></h2>
</section>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><strong>Biomedical Networks</strong></p>
<ul class="simple">
<li><p>1.1. Libraries</p></li>
<li><p>1.2. TCGA-LUAD project</p></li>
<li><p>1.3. Metadata Analysis</p></li>
<li><p>1.4. Expression Data Analysis</p></li>
</ul>
</li>
<li><p><strong>Data Filtering and Preprocessing</strong></p>
<ul class="simple">
<li><p>2.1. Thresholding Based on Gene Expression Levels</p></li>
<li><p>2.2. Gene Retention at Various Thresholds</p></li>
<li><p>2.3. Thresholding Based on Gene Expression Variance</p></li>
<li><p>2.4. Gene Retention at Different Variance Levels</p></li>
</ul>
</li>
<li><p><strong>Gene Symbol Conversion</strong></p>
<ul class="simple">
<li><p>3.1. Checking for Missing Data and Outliers</p></li>
<li><p>3.2. Addressing Data Heterogeneity</p></li>
</ul>
</li>
<li><p><strong>Correlation Matrix Calculation</strong></p></li>
<li><p><strong>Graph Construction from Correlation Matrices</strong></p></li>
<li><p><strong>Network Cleaning</strong></p></li>
<li><p><strong>Sparsification Methods for Networks</strong></p></li>
<li><p><strong>Interactive Network Visualization</strong></p></li>
<li><p><strong>Storing and Managing Generated Networks</strong></p></li>
<li><p><strong>Patient Similarity Network Construction</strong></p></li>
<li><p><strong>DNA Methylation Network Analysis</strong></p></li>
<li><p><strong>Setting up Protein Interaction Network Using STRING Database</strong></p></li>
</ol>
</section>
</section>
<section id="biomedical-networks">
<h1>1. Biomedical Networks<a class="headerlink" href="#biomedical-networks" title="Permalink to this heading">#</a></h1>
<section id="libraries">
<h2>1.1 Libraries<a class="headerlink" href="#libraries" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># standard libraries</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="c1"># scientific and data manipulation libraries</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span> <span class="nn">sklearn.feature_selection</span> <span class="kn">import</span> <span class="n">mutual_info_regression</span>
<span class="kn">from</span> <span class="nn">astropy.stats</span> <span class="kn">import</span> <span class="n">median_absolute_deviation</span>
<span class="kn">import</span> <span class="nn">mygene</span>
<span class="kn">import</span> <span class="nn">astropy</span>

<span class="c1"># graph and network libraries</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>

<span class="c1"># visualization libraries</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">plotly.io</span> <span class="k">as</span> <span class="nn">pio</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="tcga-luad-project">
<h2>1.2 TCGA-LUAD Project<a class="headerlink" href="#tcga-luad-project" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p><strong>Title</strong>: The Cancer Genome Atlas Lung Adenocarcinoma (TCGA-LUAD)</p></li>
<li><p><strong>Main Focus</strong>: Study of lung adenocarcinoma (a common type of lung cancer)</p></li>
<li><p><strong>Data Collected</strong>: Comprehensive genomic, epigenomic, transcriptomic, and proteomic data from lung adenocarcinoma samples</p></li>
<li><p><strong>Disease Types</strong>:</p>
<ul>
<li><p>Acinar Cell Neoplasms</p></li>
<li><p>Adenomas and Adenocarcinomas</p></li>
<li><p>Cystic, Mucinous, and Serous Neoplasms</p></li>
</ul>
</li>
<li><p><strong>Number of Cases</strong>: 585 (498 with transcriptomic data)</p></li>
<li><p><strong>Data Accessibility</strong>: Available on the NIH-GDC Data Portal</p></li>
<li><p><strong>Link</strong>: <a class="reference external" href="https://portal.gdc.cancer.gov/projects/TCGA-LUAD">TCGA-LUAD Project Page</a></p></li>
</ul>
<p>Our main focus in this notebook is on RNA-Seq based, already preprocessed gene expression data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># main data directory for the project</span>

<span class="n">raw_data_dir</span> <span class="o">=</span> <span class="s1">&#39;/data/raw&#39;</span>
<span class="n">intermediate_data_dir</span> <span class="o">=</span> <span class="s1">&#39;/data/intermediate&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load the data from the pickle file</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">raw_data_dir</span><span class="p">,</span><span class="s2">&quot;ISMB_TCGA_GE.pkl&quot;</span><span class="p">),</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Obviously this is not <strong>yet</strong> a graph dataset, so our first course of action is to decide how are we going to analyse this dataset. <br />
There are several ways we could do that, however today we are focusing on the network biology aspect of cancer genome data. <br />
In order to construct a biological network, we are going to first:</p>
<ul class="simple">
<li><p>examine the TCGA metadata</p></li>
<li><p>come up with useful strategies to tackle the large data size</p></li>
<li><p>create the basis of a biological network</p></li>
</ul>
</section>
<section id="metadata-analysis">
<h2>1.3 Metadata Analysis<a class="headerlink" href="#metadata-analysis" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;datExpr&#39;, &#39;datMeta&#39;])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;datMeta&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>patient</th>
      <th>race</th>
      <th>gender</th>
      <th>sample_type</th>
      <th>cigarettes_per_day</th>
      <th>Smoked</th>
      <th>sizeFactor</th>
      <th>replaceable</th>
    </tr>
    <tr>
      <th>_row</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-38-7271</th>
      <td>TCGA-38-7271</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.3699</td>
      <td>Smoker</td>
      <td>0.5841</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-55-7914</th>
      <td>TCGA-55-7914</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>0.274</td>
      <td>Smoker</td>
      <td>0.9873</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-95-7043</th>
      <td>TCGA-95-7043</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>2.1918</td>
      <td>Smoker</td>
      <td>0.5439</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>TCGA-73-4658</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.3699</td>
      <td>Smoker</td>
      <td>0.7715</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-86-8076</th>
      <td>TCGA-86-8076</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>0</td>
      <td>Never</td>
      <td>1.313</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-86-8073</th>
      <td>TCGA-86-8073</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>2.1918</td>
      <td>Smoker</td>
      <td>1.9733</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-MN-A4N4</th>
      <td>TCGA-MN-A4N4</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>0</td>
      <td>Never</td>
      <td>1.0464</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-53-7626</th>
      <td>TCGA-53-7626</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.9178</td>
      <td>Smoker</td>
      <td>1.7208</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-44-A47G</th>
      <td>TCGA-44-A47G</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.5342</td>
      <td>Smoker</td>
      <td>0.8926</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>TCGA-55-6969</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>0</td>
      <td>Never</td>
      <td>1.4732</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>498 rows Ã— 8 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;datMeta&quot;</span><span class="p">][</span><span class="s2">&quot;patient&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">size</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>498
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;datMeta&quot;</span><span class="p">][</span><span class="s1">&#39;sample_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>sample_type
Primary Tumor          469
Solid Tissue Normal     29
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set up the figure and axes for a 2-column layout</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">18</span><span class="p">,</span> <span class="mi">18</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Metadata Distributions&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Plot 1: Distribution of Race</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;race&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Race&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Race&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="c1"># Plot 2: Gender Distribution</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;gender&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;magma&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Gender Distribution&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Gender&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>

<span class="c1"># Plot 3: Sample Type Distribution</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;sample_type&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Sample Type Distribution&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Sample Type&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="c1"># Plot 4: Distribution of Cigarettes Per Day</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">][</span><span class="s1">&#39;cigarettes_per_day&#39;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Cigarettes Per Day&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Cigarettes Per Day&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>

<span class="c1"># Plot 5: Smoking Status by Gender</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Smoked&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;gender&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Smoking Status by Gender&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Smoking Status&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Gender&#39;</span><span class="p">)</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/5ce11f62a0e9f129a0d0026ba63ebf236b902ab58471d9f4bc1ec6921907dbd7.png" src="../_images/5ce11f62a0e9f129a0d0026ba63ebf236b902ab58471d9f4bc1ec6921907dbd7.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">][</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">][</span><span class="s1">&#39;sample_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Primary Tumor&#39;</span><span class="p">][</span><span class="s1">&#39;Smoked&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Smoked
Smoker    323
Never     146
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="expression-data-analysis">
<h1>Expression data analysis<a class="headerlink" href="#expression-data-analysis" title="Permalink to this heading">#</a></h1>
<p>This dataset contains gene expression levels for various samples, identified by their TCGA (The Cancer Genome Atlas) codes.<br />
Each row represents a different sample, while each column represents a different gene, identified by its Ensembl gene ID.<br />
The values in the table are the expression levels of the genes for each sample.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># let&#39;s create a new variable for the expression data alone, just for ease of use, and then inspect it</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;datExpr&quot;</span><span class="p">]</span>
<span class="n">expression_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>_row</th>
      <th>ENSG00000000003.15</th>
      <th>ENSG00000000419.13</th>
      <th>ENSG00000000457.14</th>
      <th>ENSG00000000460.17</th>
      <th>ENSG00000000938.13</th>
      <th>ENSG00000000971.16</th>
      <th>ENSG00000001036.14</th>
      <th>ENSG00000001084.13</th>
      <th>ENSG00000001167.14</th>
      <th>ENSG00000001460.18</th>
      <th>...</th>
      <th>ENSG00000288586.1</th>
      <th>ENSG00000288596.2</th>
      <th>ENSG00000288598.1</th>
      <th>ENSG00000288611.1</th>
      <th>ENSG00000288612.1</th>
      <th>ENSG00000288638.1</th>
      <th>ENSG00000288658.1</th>
      <th>ENSG00000288663.1</th>
      <th>ENSG00000288670.1</th>
      <th>ENSG00000288675.1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-38-7271</th>
      <td>11.3668</td>
      <td>10.3673</td>
      <td>9.7884</td>
      <td>8.2552</td>
      <td>10.6173</td>
      <td>12.8447</td>
      <td>10.8836</td>
      <td>10.4437</td>
      <td>10.8112</td>
      <td>9.1178</td>
      <td>...</td>
      <td>4.4411</td>
      <td>5.6909</td>
      <td>3.8172</td>
      <td>5.5378</td>
      <td>4.5573</td>
      <td>5.4255</td>
      <td>5.1694</td>
      <td>4.0073</td>
      <td>7.97</td>
      <td>4.664</td>
    </tr>
    <tr>
      <th>TCGA-55-7914</th>
      <td>11.5434</td>
      <td>10.5282</td>
      <td>9.7292</td>
      <td>7.9951</td>
      <td>8.4858</td>
      <td>10.4696</td>
      <td>10.4737</td>
      <td>7.9065</td>
      <td>11.8983</td>
      <td>8.5842</td>
      <td>...</td>
      <td>4.768</td>
      <td>8.0948</td>
      <td>4.2182</td>
      <td>3.5233</td>
      <td>4.975</td>
      <td>2.4821</td>
      <td>6.0753</td>
      <td>4.975</td>
      <td>8.8543</td>
      <td>4.3015</td>
    </tr>
    <tr>
      <th>TCGA-95-7043</th>
      <td>11.411</td>
      <td>11.2018</td>
      <td>9.4449</td>
      <td>8.3546</td>
      <td>7.3211</td>
      <td>10.7856</td>
      <td>11.6185</td>
      <td>14.1836</td>
      <td>10.8329</td>
      <td>8.8155</td>
      <td>...</td>
      <td>4.2248</td>
      <td>7.3539</td>
      <td>5.0121</td>
      <td>4.2248</td>
      <td>4.8311</td>
      <td>4.7304</td>
      <td>5.5053</td>
      <td>4.9246</td>
      <td>7.6064</td>
      <td>3.8623</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>12.2149</td>
      <td>10.3249</td>
      <td>9.234</td>
      <td>7.7537</td>
      <td>11.0656</td>
      <td>12.3631</td>
      <td>10.9023</td>
      <td>10.06</td>
      <td>11.2092</td>
      <td>9.0663</td>
      <td>...</td>
      <td>4.2146</td>
      <td>6.0253</td>
      <td>3.8231</td>
      <td>4.0984</td>
      <td>4.6727</td>
      <td>7.1082</td>
      <td>7.1615</td>
      <td>3.6533</td>
      <td>7.3223</td>
      <td>5.2312</td>
    </tr>
    <tr>
      <th>TCGA-86-8076</th>
      <td>11.2882</td>
      <td>10.2095</td>
      <td>9.8186</td>
      <td>7.9844</td>
      <td>10.5213</td>
      <td>13.7848</td>
      <td>11.2318</td>
      <td>10.8315</td>
      <td>10.5784</td>
      <td>9.6428</td>
      <td>...</td>
      <td>4.7844</td>
      <td>6.8675</td>
      <td>4.5078</td>
      <td>2.4821</td>
      <td>5.1468</td>
      <td>3.3897</td>
      <td>4.2833</td>
      <td>4.4553</td>
      <td>8.1011</td>
      <td>4.8254</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-86-8073</th>
      <td>10.7676</td>
      <td>10.5503</td>
      <td>9.0277</td>
      <td>8.2811</td>
      <td>9.9471</td>
      <td>11.8641</td>
      <td>11.237</td>
      <td>14.2843</td>
      <td>10.9497</td>
      <td>8.8266</td>
      <td>...</td>
      <td>4.2611</td>
      <td>7.157</td>
      <td>4.2187</td>
      <td>5.2566</td>
      <td>5.276</td>
      <td>2.4821</td>
      <td>6.695</td>
      <td>5.1967</td>
      <td>7.8098</td>
      <td>4.3418</td>
    </tr>
    <tr>
      <th>TCGA-MN-A4N4</th>
      <td>11.656</td>
      <td>11.0739</td>
      <td>9.5578</td>
      <td>8.7883</td>
      <td>8.9532</td>
      <td>11.0791</td>
      <td>10.5202</td>
      <td>11.331</td>
      <td>11.3659</td>
      <td>8.8701</td>
      <td>...</td>
      <td>4.3317</td>
      <td>7.6037</td>
      <td>5.0518</td>
      <td>7.4063</td>
      <td>6.5835</td>
      <td>3.0746</td>
      <td>5.9828</td>
      <td>4.9634</td>
      <td>7.9208</td>
      <td>5.1745</td>
    </tr>
    <tr>
      <th>TCGA-53-7626</th>
      <td>10.8912</td>
      <td>10.1742</td>
      <td>9.6307</td>
      <td>8.0179</td>
      <td>10.6328</td>
      <td>11.247</td>
      <td>10.4815</td>
      <td>9.8428</td>
      <td>10.5795</td>
      <td>9.86</td>
      <td>...</td>
      <td>5.6458</td>
      <td>6.6984</td>
      <td>4.646</td>
      <td>3.1345</td>
      <td>4.8113</td>
      <td>3.8962</td>
      <td>5.9768</td>
      <td>4.2816</td>
      <td>8.4192</td>
      <td>4.4566</td>
    </tr>
    <tr>
      <th>TCGA-44-A47G</th>
      <td>11.32</td>
      <td>9.9137</td>
      <td>9.3312</td>
      <td>7.8703</td>
      <td>11.3029</td>
      <td>12.8687</td>
      <td>11.2659</td>
      <td>12.4677</td>
      <td>10.9404</td>
      <td>9.4188</td>
      <td>...</td>
      <td>4.6125</td>
      <td>8.0939</td>
      <td>5.1263</td>
      <td>2.4821</td>
      <td>5.2619</td>
      <td>5.0279</td>
      <td>8.793</td>
      <td>5.173</td>
      <td>7.951</td>
      <td>4.465</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>10.5764</td>
      <td>10.9022</td>
      <td>9.3425</td>
      <td>8.8147</td>
      <td>9.6115</td>
      <td>12.3155</td>
      <td>11.216</td>
      <td>11.1273</td>
      <td>10.311</td>
      <td>9.123</td>
      <td>...</td>
      <td>5.9603</td>
      <td>7.2623</td>
      <td>4.5931</td>
      <td>4.1326</td>
      <td>4.5495</td>
      <td>3.1863</td>
      <td>3.9293</td>
      <td>4.828</td>
      <td>7.5868</td>
      <td>4.1326</td>
    </tr>
  </tbody>
</table>
<p>498 rows Ã— 22637 columns</p>
</div></div></div>
</div>
<p>498 rows Ã— 22637 columns is a quite large matrix, thus we have to consider reducing the dimensionality.</p>
</section>
<section id="data-filtering-and-preprocessing">
<h1>2. Data Filtering and Preprocessing<a class="headerlink" href="#data-filtering-and-preprocessing" title="Permalink to this heading">#</a></h1>
<p>Ideas?</p>
<ul class="simple">
<li><p>Filter by Mean Expression: Select genes with high mean expression levels</p></li>
<li><p>Filter by Variance: Select genes with high variance across samples, as low variance genes might not contribute significantly to the analysis.</p></li>
</ul>
<p>We are doing this to make sure our computations are</p>
<ul class="simple">
<li><p>computationally efficient,</p></li>
<li><p>the network complexity is manageable,</p></li>
<li><p>the biological signal is enhanced thus we make sure our analysis is biologically relevant.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># A few preliminary steps that might be useful for data cleaning and preprocessing</span>
<span class="c1"># Ensure all columns are numeric</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s1">&#39;coerce&#39;</span><span class="p">)</span>

<span class="c1"># Drop columns that could not be converted to numeric (if any)</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># In case we want to check the shape of the data further down the line</span>
<span class="n">expression_data</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(498, 22637)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Checking for duplicate rows and columns</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of duplicate indices: </span><span class="si">{</span><span class="n">expression_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of duplicate columns: </span><span class="si">{</span><span class="n">expression_data</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of duplicate indices: 0
Number of duplicate columns: 0
</pre></div>
</div>
</div>
</div>
<section id="analysing-gene-expression-levels">
<h2>2.1 Analysing Gene Expression Levels<a class="headerlink" href="#analysing-gene-expression-levels" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the distribution of gene expression levels </span>

<span class="c1"># Calculate the mean expression level for each gene</span>
<span class="n">gene_means</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Plot the distribution of gene expression levels</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">gene_means</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mean Gene Expression&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Gene Expression Levels&#39;</span><span class="p">)</span>

<span class="c1"># mean of the gene means</span>
<span class="n">threshold</span> <span class="o">=</span> <span class="n">gene_means</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Mean = </span><span class="si">{</span><span class="n">threshold</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/c1159eb425082c8c00119a8a0dccff11cda1037b86d5414e2fe821e1ccce3e73.png" src="../_images/c1159eb425082c8c00119a8a0dccff11cda1037b86d5414e2fe821e1ccce3e73.png" />
</div>
</div>
<p>The histogram shows the frequency of genes at different mean expression levels.<br />
Dashed red line shows the mean of the mean gene expression.  (8.11)<br />
The histogram show a bimodal distribution - does this mean two groups of genes?<br />
Some genes are consistently expressed at lower levels (housekeeping genes?), while others are expressed at higher levels.</p>
<p>We could use this for thresholding, however it is not informative about the variability of the genes.</p>
<p>Still, letâ€™s inspect how many genes would retain in our dataset at different expression thresholds.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to filter out low-expressed genes</span>
<span class="k">def</span> <span class="nf">filter_low_expression_genes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
    <span class="c1"># Calculate the mean expression for each gene</span>
    <span class="n">gene_means</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Filter out genes with mean expression below the threshold</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">gene_means</span> <span class="o">&gt;=</span> <span class="n">threshold</span>
    <span class="n">filtered_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">mask</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">filtered_data</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Gene Retention at Various Thresholds</span>
<span class="c1"># Define a range of thresholds</span>
<span class="n">thresholds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># List to store the number of genes retained at each threshold</span>
<span class="n">num_genes</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Assuming df_renamed is your DataFrame with gene expression data</span>
<span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">filter_low_expression_genes</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="n">threshold</span><span class="p">)</span>
    <span class="n">num_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_filtered</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Plot the results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">thresholds</span><span class="p">,</span> <span class="n">num_genes</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Threshold&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained at Different Thresholds&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/f0dba6767841754602fe39e1c2102093b5452542f746036fcd80e22dd5064acb.png" src="../_images/f0dba6767841754602fe39e1c2102093b5452542f746036fcd80e22dd5064acb.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign the filtered data to a new variable</span>
<span class="c1"># df_filtered_expression = filter_low_expression_genes(expression_data, threshold=9)</span>
<span class="c1"># print(f&quot;Filtered data shape: {df_filtered_expression.shape}&quot;)  # Check the new shape to confirm filtering</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="thresholding-based-on-gene-expression-variance">
<h2>2.3 Thresholding Based on Gene Expression Variance<a class="headerlink" href="#thresholding-based-on-gene-expression-variance" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>High-dimensional data can be computationally expensive to process.</p></li>
<li><p>Genes with higher variance across samples are likely to be more biologically relevant.</p></li>
<li><p>Low-variance genes often represent housekeeping genes or noise.</p></li>
</ol>
<section id="task-1-variance-based-gene-filtering">
<h3>Task 1: Variance-Based Gene Filtering<a class="headerlink" href="#task-1-variance-based-gene-filtering" title="Permalink to this heading">#</a></h3>
<p>Write a function that takes in a dataframe and a threshold value as argument, and call it filter_high_variance_genes()</p>
<section id="step-1-calculate-variance">
<h4>Step 1: Calculate Variance<a class="headerlink" href="#step-1-calculate-variance" title="Permalink to this heading">#</a></h4>
<p>Compute the variance of expression levels for each gene across all samples.</p>
</section>
<section id="step-2-set-a-variance-threshold">
<h4>Step 2: Set a Variance Threshold<a class="headerlink" href="#step-2-set-a-variance-threshold" title="Permalink to this heading">#</a></h4>
<p>Determine a threshold for variance. Genes with variance below this threshold are considered to have low variance and can be filtered out.</p>
</section>
<section id="step-3-filter-genes">
<h4>Step 3: Filter Genes<a class="headerlink" href="#step-3-filter-genes" title="Permalink to this heading">#</a></h4>
<p>Retain only those genes with variance above the threshold for network construction.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">filter_high_variance_genes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters out genes (columns) with variance below the specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    data (pd.DataFrame): DataFrame with gene expression data where rows are samples and columns are genes.</span>
<span class="sd">    threshold (float): Minimum variance level to keep a gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">    pd.DataFrame: Filtered DataFrame with genes having variance above the threshold.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate the variance for each gene (column)</span>
    <span class="n">gene_variances</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Create a boolean mask to filter out genes with variance below the threshold</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">gene_variances</span> <span class="o">&gt;=</span> <span class="n">threshold</span>
    <span class="c1"># Apply the mask to filter the DataFrame</span>
    <span class="n">filtered_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">mask</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">filtered_data</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="gene-retention-at-different-variance-levels">
<h2>2.4 Gene Retention at Different Variance Levels<a class="headerlink" href="#gene-retention-at-different-variance-levels" title="Permalink to this heading">#</a></h2>
<p>Idea:</p>
<p>Visualise the gene retention at different thresholds.</p>
<p>Calculate the 75th percentile of the variance distribution, and use it as a threshold.<br />
It is a good balance between retaining enough data for meaningful analysis and removing low-variance noise</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a range of variance thresholds</span>
<span class="n">variance_thresholds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># List to store the number of genes retained at each threshold</span>
<span class="n">num_genes</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Assuming df_renamed is your DataFrame with gene expression data</span>
<span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">variance_thresholds</span><span class="p">:</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">filter_high_variance_genes</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="n">threshold</span><span class="p">)</span>
    <span class="n">num_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_filtered</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Calculate the variance for each gene</span>
<span class="n">gene_variances</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="c1"># Calculate the 75th percentile of the variance distribution</span>
<span class="n">variance_threshold</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">gene_variances</span><span class="p">,</span> <span class="mi">75</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Chosen Variance Threshold: </span><span class="si">{</span><span class="n">variance_threshold</span><span class="w"> </span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Plot the results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">variance_thresholds</span><span class="p">,</span> <span class="n">num_genes</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">variance_threshold</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;75th Percentile Threshold = </span><span class="si">{</span><span class="n">variance_threshold</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Variance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained at Different Variance Thresholds&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Chosen Variance Threshold: 1.2018901296749187
</pre></div>
</div>
<img alt="../_images/f47d742f7f0342b89a98144cef054c8917eb21e5f28f22e59394a4766e675de6.png" src="../_images/f47d742f7f0342b89a98144cef054c8917eb21e5f28f22e59394a4766e675de6.png" />
</div>
</div>
<p>Letâ€™s assign this value as threshold to focus on genes with higher variance.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign the filtered data to a new variable</span>
<span class="c1"># Focusing on the top quartile (75th percentile) of genes with the highest variance</span>
<span class="n">df_filtered_variance</span> <span class="o">=</span> <span class="n">filter_high_variance_genes</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mf">1.2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtered data shape: </span><span class="si">{</span><span class="n">df_filtered_variance</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># Check the new shape to confirm filtering</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Filtered data shape: (498, 5671)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># df_filtered_variance.describe()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="gene-symbol-conversion-ensg-to-hgcn">
<h1>3. Gene Symbol Conversion: ENSG to HGCN<a class="headerlink" href="#gene-symbol-conversion-ensg-to-hgcn" title="Permalink to this heading">#</a></h1>
<p>Converting Ensembl gene IDs (ENSG) to HGNC (HUGO Gene Nomenclature Committee) gene symbols is often a good practice as HGCN is</p>
<ul class="simple">
<li><p>standard identifier,</p></li>
<li><p>more human readable,</p></li>
<li><p>is widely used in publications, thus helps domain experts understand more effectively the analysis.</p></li>
</ul>
<p>We are using for this conversion the <code class="docutils literal notranslate"><span class="pre">mygene</span></code> package. <a class="reference external" href="https://docs.mygene.info/projects/mygene-py/en/latest/">DOC</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">rename_ensembl_to_gene_names</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">chunk_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Renames Ensembl gene IDs to gene names using mygene.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    df (pd.DataFrame): DataFrame with Ensembl gene IDs as columns.</span>
<span class="sd">    chunk_size (int): Number of Ensembl IDs to query at a time.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    pd.DataFrame: DataFrame with gene names as columns, excluding genes that couldn&#39;t be mapped.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Make a copy of the DataFrame to avoid modifying the original</span>
    <span class="n">df_copy</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Remove the `.number` suffix from ENSG IDs</span>
    <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Initialize mygene client</span>
    <span class="n">mg</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">MyGeneInfo</span><span class="p">()</span>

    <span class="c1"># Split ENSG IDs into smaller chunks</span>
    <span class="k">def</span> <span class="nf">chunks</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Yield successive n-sized chunks from lst.&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">lst</span><span class="p">),</span> <span class="n">n</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">lst</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">n</span><span class="p">]</span>

    <span class="n">ensg_ids</span> <span class="o">=</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">gene_mappings</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">chunks</span><span class="p">(</span><span class="n">ensg_ids</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">):</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;ensembl.gene&#39;</span><span class="p">,</span> <span class="n">fields</span><span class="o">=</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">)</span>
        <span class="n">gene_mappings</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

    <span class="c1"># Create a mapping from ENSG to gene symbol, handle missing mappings</span>
    <span class="n">ensg_to_gene</span> <span class="o">=</span> <span class="p">{</span><span class="n">item</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]:</span> <span class="n">item</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_mappings</span><span class="p">}</span>
    
    <span class="c1"># Log the unmapped genes</span>
    <span class="n">unmapped_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">ensg_ids</span> <span class="k">if</span> <span class="n">ensg_to_gene</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">unmapped_genes</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unmapped genes: </span><span class="si">{</span><span class="n">unmapped_genes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Filter out genes that couldn&#39;t be mapped</span>
    <span class="n">mapped_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="k">for</span> <span class="n">ensg</span> <span class="ow">in</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>

    <span class="c1"># Filter the DataFrame to only include columns that have been mapped</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">ensg_to_gene</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>

    <span class="c1"># Further filter to ensure we have the same number of columns as mapped gene names</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="n">ensg</span> <span class="k">for</span> <span class="n">ensg</span> <span class="ow">in</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]]</span>

    <span class="c1"># Assign new column names</span>
    <span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="k">for</span> <span class="n">ensg</span> <span class="ow">in</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

    <span class="c1"># Handle duplicate gene names by aggregating them (e.g., by taking the mean)</span>
    <span class="n">df_final</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>

    <span class="k">return</span> <span class="n">df_final</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># convert the ensembl gene IDs to gene names</span>
<span class="n">df_renamed</span> <span class="o">=</span> <span class="n">rename_ensembl_to_gene_names</span><span class="p">(</span><span class="n">df_filtered_variance</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1 input query terms found no hit:	[&#39;ENSG00000168078&#39;]
1 input query terms found no hit:	[&#39;ENSG00000180525&#39;]
2 input query terms found no hit:	[&#39;ENSG00000236740&#39;, &#39;ENSG00000239467&#39;]
9 input query terms found no hit:	[&#39;ENSG00000256618&#39;, &#39;ENSG00000259820&#39;, &#39;ENSG00000259834&#39;, &#39;ENSG00000261068&#39;, &#39;ENSG00000261534&#39;, &#39;ENS
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Unmapped genes: [&#39;ENSG00000166104&#39;, &#39;ENSG00000168078&#39;, &#39;ENSG00000180525&#39;, &#39;ENSG00000189229&#39;, &#39;ENSG00000197476&#39;, &#39;ENSG00000203635&#39;, &#39;ENSG00000204584&#39;, &#39;ENSG00000213058&#39;, &#39;ENSG00000214776&#39;, &#39;ENSG00000214797&#39;, &#39;ENSG00000215146&#39;, &#39;ENSG00000217801&#39;, &#39;ENSG00000218048&#39;, &#39;ENSG00000218175&#39;, &#39;ENSG00000218274&#39;, &#39;ENSG00000218617&#39;, &#39;ENSG00000220154&#39;, &#39;ENSG00000222000&#39;, &#39;ENSG00000223829&#39;, &#39;ENSG00000223931&#39;, &#39;ENSG00000224251&#39;, &#39;ENSG00000224585&#39;, &#39;ENSG00000224945&#39;, &#39;ENSG00000226025&#39;, &#39;ENSG00000226286&#39;, &#39;ENSG00000226468&#39;, &#39;ENSG00000226668&#39;, &#39;ENSG00000226816&#39;, &#39;ENSG00000227619&#39;, &#39;ENSG00000227755&#39;, &#39;ENSG00000228113&#39;, &#39;ENSG00000228477&#39;, &#39;ENSG00000228509&#39;, &#39;ENSG00000228613&#39;, &#39;ENSG00000229312&#39;, &#39;ENSG00000229896&#39;, &#39;ENSG00000229970&#39;, &#39;ENSG00000230623&#39;, &#39;ENSG00000230628&#39;, &#39;ENSG00000230699&#39;, &#39;ENSG00000230882&#39;, &#39;ENSG00000231331&#39;, &#39;ENSG00000231533&#39;, &#39;ENSG00000231703&#39;, &#39;ENSG00000231881&#39;, &#39;ENSG00000231892&#39;, &#39;ENSG00000231990&#39;, &#39;ENSG00000232006&#39;, &#39;ENSG00000232949&#39;, &#39;ENSG00000234156&#39;, &#39;ENSG00000234648&#39;, &#39;ENSG00000235563&#39;, &#39;ENSG00000235584&#39;, &#39;ENSG00000235721&#39;, &#39;ENSG00000235939&#39;, &#39;ENSG00000235994&#39;, &#39;ENSG00000236212&#39;, &#39;ENSG00000236283&#39;, &#39;ENSG00000236449&#39;, &#39;ENSG00000236740&#39;, &#39;ENSG00000237550&#39;, &#39;ENSG00000237813&#39;, &#39;ENSG00000238279&#39;, &#39;ENSG00000239467&#39;, &#39;ENSG00000244306&#39;, &#39;ENSG00000247134&#39;, &#39;ENSG00000248347&#39;, &#39;ENSG00000248545&#39;, &#39;ENSG00000249001&#39;, &#39;ENSG00000249007&#39;, &#39;ENSG00000249199&#39;, &#39;ENSG00000249743&#39;, &#39;ENSG00000249746&#39;, &#39;ENSG00000250049&#39;, &#39;ENSG00000250509&#39;, &#39;ENSG00000250770&#39;, &#39;ENSG00000250978&#39;, &#39;ENSG00000251095&#39;, &#39;ENSG00000251532&#39;, &#39;ENSG00000253217&#39;, &#39;ENSG00000253227&#39;, &#39;ENSG00000253288&#39;, &#39;ENSG00000253339&#39;, &#39;ENSG00000253361&#39;, &#39;ENSG00000254054&#39;, &#39;ENSG00000254143&#39;, &#39;ENSG00000254290&#39;, &#39;ENSG00000254367&#39;, &#39;ENSG00000254632&#39;, &#39;ENSG00000254851&#39;, &#39;ENSG00000254988&#39;, &#39;ENSG00000255026&#39;, &#39;ENSG00000255085&#39;, &#39;ENSG00000255182&#39;, &#39;ENSG00000255310&#39;, &#39;ENSG00000255345&#39;, &#39;ENSG00000255366&#39;, &#39;ENSG00000255462&#39;, &#39;ENSG00000255847&#39;, &#39;ENSG00000255910&#39;, &#39;ENSG00000255983&#39;, &#39;ENSG00000256417&#39;, &#39;ENSG00000256443&#39;, &#39;ENSG00000256615&#39;, &#39;ENSG00000256618&#39;, &#39;ENSG00000256673&#39;, &#39;ENSG00000256969&#39;, &#39;ENSG00000257474&#39;, &#39;ENSG00000257512&#39;, &#39;ENSG00000257604&#39;, &#39;ENSG00000257826&#39;, &#39;ENSG00000257829&#39;, &#39;ENSG00000257894&#39;, &#39;ENSG00000258053&#39;, &#39;ENSG00000258077&#39;, &#39;ENSG00000258168&#39;, &#39;ENSG00000258342&#39;, &#39;ENSG00000258346&#39;, &#39;ENSG00000258661&#39;, &#39;ENSG00000258676&#39;, &#39;ENSG00000258752&#39;, &#39;ENSG00000258837&#39;, &#39;ENSG00000258844&#39;, &#39;ENSG00000259087&#39;, &#39;ENSG00000259094&#39;, &#39;ENSG00000259342&#39;, &#39;ENSG00000259345&#39;, &#39;ENSG00000259353&#39;, &#39;ENSG00000259425&#39;, &#39;ENSG00000259723&#39;, &#39;ENSG00000259820&#39;, &#39;ENSG00000259834&#39;, &#39;ENSG00000259925&#39;, &#39;ENSG00000260197&#39;, &#39;ENSG00000260289&#39;, &#39;ENSG00000260328&#39;, &#39;ENSG00000260337&#39;, &#39;ENSG00000260430&#39;, &#39;ENSG00000260597&#39;, &#39;ENSG00000260604&#39;, &#39;ENSG00000260710&#39;, &#39;ENSG00000260711&#39;, &#39;ENSG00000260877&#39;, &#39;ENSG00000260997&#39;, &#39;ENSG00000261068&#39;, &#39;ENSG00000261104&#39;, &#39;ENSG00000261123&#39;, &#39;ENSG00000261172&#39;, &#39;ENSG00000261296&#39;, &#39;ENSG00000261305&#39;, &#39;ENSG00000261502&#39;, &#39;ENSG00000261534&#39;, &#39;ENSG00000261578&#39;, &#39;ENSG00000261600&#39;, &#39;ENSG00000261762&#39;, &#39;ENSG00000261786&#39;, &#39;ENSG00000261804&#39;, &#39;ENSG00000262539&#39;, &#39;ENSG00000262580&#39;, &#39;ENSG00000262714&#39;, &#39;ENSG00000262873&#39;, &#39;ENSG00000262877&#39;, &#39;ENSG00000262920&#39;, &#39;ENSG00000262979&#39;, &#39;ENSG00000263301&#39;, &#39;ENSG00000263893&#39;, &#39;ENSG00000264112&#39;, &#39;ENSG00000264659&#39;, &#39;ENSG00000265298&#39;, &#39;ENSG00000265750&#39;, &#39;ENSG00000266302&#39;, &#39;ENSG00000266830&#39;, &#39;ENSG00000266869&#39;, &#39;ENSG00000266968&#39;, &#39;ENSG00000267072&#39;, &#39;ENSG00000267383&#39;, &#39;ENSG00000267466&#39;, &#39;ENSG00000267506&#39;, &#39;ENSG00000267640&#39;, &#39;ENSG00000267886&#39;, &#39;ENSG00000267892&#39;, &#39;ENSG00000268362&#39;, &#39;ENSG00000268416&#39;, &#39;ENSG00000268555&#39;, &#39;ENSG00000268833&#39;, &#39;ENSG00000268903&#39;, &#39;ENSG00000269028&#39;, &#39;ENSG00000269353&#39;, &#39;ENSG00000269918&#39;, &#39;ENSG00000269954&#39;, &#39;ENSG00000269981&#39;, &#39;ENSG00000270460&#39;, &#39;ENSG00000270605&#39;, &#39;ENSG00000270933&#39;, &#39;ENSG00000270977&#39;, &#39;ENSG00000271392&#39;, &#39;ENSG00000271959&#39;, &#39;ENSG00000272079&#39;, &#39;ENSG00000272141&#39;, &#39;ENSG00000272279&#39;, &#39;ENSG00000272379&#39;, &#39;ENSG00000272405&#39;, &#39;ENSG00000272502&#39;, &#39;ENSG00000272512&#39;, &#39;ENSG00000272622&#39;, &#39;ENSG00000272626&#39;, &#39;ENSG00000272720&#39;, &#39;ENSG00000272825&#39;, &#39;ENSG00000272865&#39;, &#39;ENSG00000273108&#39;, &#39;ENSG00000273151&#39;, &#39;ENSG00000273179&#39;, &#39;ENSG00000273203&#39;, &#39;ENSG00000273230&#39;, &#39;ENSG00000273295&#39;, &#39;ENSG00000273301&#39;, &#39;ENSG00000273328&#39;, &#39;ENSG00000273760&#39;, &#39;ENSG00000273906&#39;, &#39;ENSG00000274002&#39;, &#39;ENSG00000274414&#39;, &#39;ENSG00000274979&#39;, &#39;ENSG00000275678&#39;, &#39;ENSG00000276093&#39;, &#39;ENSG00000276851&#39;, &#39;ENSG00000276980&#39;, &#39;ENSG00000277287&#39;, &#39;ENSG00000277840&#39;, &#39;ENSG00000278263&#39;, &#39;ENSG00000278266&#39;, &#39;ENSG00000278484&#39;, &#39;ENSG00000278864&#39;, &#39;ENSG00000278912&#39;, &#39;ENSG00000278962&#39;, &#39;ENSG00000279017&#39;, &#39;ENSG00000279286&#39;, &#39;ENSG00000279511&#39;, &#39;ENSG00000279518&#39;, &#39;ENSG00000279662&#39;, &#39;ENSG00000279806&#39;, &#39;ENSG00000279878&#39;, &#39;ENSG00000279930&#39;, &#39;ENSG00000280002&#39;, &#39;ENSG00000280027&#39;, &#39;ENSG00000280032&#39;, &#39;ENSG00000280079&#39;, &#39;ENSG00000280138&#39;, &#39;ENSG00000280202&#39;, &#39;ENSG00000280217&#39;, &#39;ENSG00000280285&#39;, &#39;ENSG00000280334&#39;, &#39;ENSG00000280399&#39;, &#39;ENSG00000280435&#39;, &#39;ENSG00000283098&#39;, &#39;ENSG00000283142&#39;, &#39;ENSG00000283413&#39;, &#39;ENSG00000283486&#39;, &#39;ENSG00000283631&#39;, &#39;ENSG00000283633&#39;, &#39;ENSG00000283907&#39;, &#39;ENSG00000284636&#39;, &#39;ENSG00000284928&#39;, &#39;ENSG00000285108&#39;, &#39;ENSG00000285128&#39;, &#39;ENSG00000285417&#39;, &#39;ENSG00000285476&#39;, &#39;ENSG00000285517&#39;, &#39;ENSG00000285601&#39;, &#39;ENSG00000285820&#39;, &#39;ENSG00000285847&#39;, &#39;ENSG00000285933&#39;, &#39;ENSG00000286009&#39;, &#39;ENSG00000286015&#39;, &#39;ENSG00000286073&#39;, &#39;ENSG00000286133&#39;, &#39;ENSG00000286191&#39;, &#39;ENSG00000286256&#39;, &#39;ENSG00000286299&#39;, &#39;ENSG00000286342&#39;, &#39;ENSG00000286355&#39;, &#39;ENSG00000286358&#39;, &#39;ENSG00000286403&#39;, &#39;ENSG00000286449&#39;, &#39;ENSG00000286456&#39;, &#39;ENSG00000286519&#39;, &#39;ENSG00000286587&#39;, &#39;ENSG00000286722&#39;, &#39;ENSG00000286750&#39;, &#39;ENSG00000286797&#39;, &#39;ENSG00000286802&#39;, &#39;ENSG00000286818&#39;, &#39;ENSG00000286830&#39;, &#39;ENSG00000286892&#39;, &#39;ENSG00000286944&#39;, &#39;ENSG00000286989&#39;, &#39;ENSG00000287003&#39;, &#39;ENSG00000287051&#39;, &#39;ENSG00000287052&#39;, &#39;ENSG00000287055&#39;, &#39;ENSG00000287059&#39;, &#39;ENSG00000287097&#39;, &#39;ENSG00000287185&#39;, &#39;ENSG00000287238&#39;, &#39;ENSG00000287245&#39;, &#39;ENSG00000287252&#39;, &#39;ENSG00000287271&#39;, &#39;ENSG00000287338&#39;, &#39;ENSG00000287352&#39;, &#39;ENSG00000287390&#39;, &#39;ENSG00000287426&#39;, &#39;ENSG00000287430&#39;, &#39;ENSG00000287460&#39;, &#39;ENSG00000287550&#39;, &#39;ENSG00000287626&#39;, &#39;ENSG00000287627&#39;, &#39;ENSG00000287656&#39;, &#39;ENSG00000287673&#39;, &#39;ENSG00000287725&#39;, &#39;ENSG00000287770&#39;, &#39;ENSG00000287778&#39;, &#39;ENSG00000287804&#39;, &#39;ENSG00000287839&#39;, &#39;ENSG00000287860&#39;, &#39;ENSG00000287879&#39;, &#39;ENSG00000287887&#39;, &#39;ENSG00000287906&#39;, &#39;ENSG00000287920&#39;, &#39;ENSG00000287929&#39;, &#39;ENSG00000288075&#39;, &#39;ENSG00000288218&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s check the new shape of the data after renaming</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Shape of df_filtered_variance:&quot;</span><span class="p">,</span> <span class="n">df_filtered_variance</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Shape of df_renamed:&quot;</span><span class="p">,</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Shape of df_filtered_variance: (498, 5671)
Shape of df_renamed: (498, 5338)
</pre></div>
</div>
</div>
</div>
<p>What could this mean?</p>
<p>The output of the <code class="docutils literal notranslate"><span class="pre">mygene</span></code> and the different number of rows indicate that some of the genes werent mapped to a gene ID or multiple were mapped to the same gene ID.<br />
In the first case we disregard those genes, in the second we calculate the mean expression value.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># let&#39;s inspect the first few rows of the renamed DataFrame</span>
<span class="n">df_renamed</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A2M</th>
      <th>A2ML1</th>
      <th>A4GALT</th>
      <th>AACSP1</th>
      <th>AADAC</th>
      <th>AADACP1</th>
      <th>AARD</th>
      <th>AASS</th>
      <th>AATBC</th>
      <th>AATK</th>
      <th>...</th>
      <th>ZNF90</th>
      <th>ZP3</th>
      <th>ZPLD1</th>
      <th>ZPLD2P</th>
      <th>ZSCAN1</th>
      <th>ZSCAN23</th>
      <th>ZSCAN31</th>
      <th>ZSCAN4</th>
      <th>ZWINT</th>
      <th>ZYG11A</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-38-7271</th>
      <td>17.7492</td>
      <td>4.4411</td>
      <td>10.1862</td>
      <td>3.5845</td>
      <td>9.6415</td>
      <td>3.5845</td>
      <td>8.0089</td>
      <td>11.2890</td>
      <td>6.8710</td>
      <td>8.6552</td>
      <td>...</td>
      <td>6.1774</td>
      <td>6.9127</td>
      <td>4.9403</td>
      <td>3.5845</td>
      <td>3.8172</td>
      <td>5.7385</td>
      <td>9.0817</td>
      <td>3.2709</td>
      <td>9.2615</td>
      <td>4.0073</td>
    </tr>
    <tr>
      <th>TCGA-55-7914</th>
      <td>14.8513</td>
      <td>4.4530</td>
      <td>8.9312</td>
      <td>3.6762</td>
      <td>6.5685</td>
      <td>3.6762</td>
      <td>6.8983</td>
      <td>10.5920</td>
      <td>7.9894</td>
      <td>8.2660</td>
      <td>...</td>
      <td>9.9162</td>
      <td>7.9426</td>
      <td>2.4821</td>
      <td>2.4821</td>
      <td>5.0222</td>
      <td>4.3796</td>
      <td>10.9565</td>
      <td>3.0918</td>
      <td>10.1334</td>
      <td>8.3219</td>
    </tr>
    <tr>
      <th>TCGA-95-7043</th>
      <td>14.1691</td>
      <td>4.5026</td>
      <td>9.0834</td>
      <td>3.8623</td>
      <td>5.5634</td>
      <td>8.0183</td>
      <td>3.6226</td>
      <td>10.5600</td>
      <td>7.5223</td>
      <td>8.8214</td>
      <td>...</td>
      <td>5.3150</td>
      <td>5.8224</td>
      <td>3.6226</td>
      <td>3.2988</td>
      <td>3.8623</td>
      <td>5.6728</td>
      <td>10.8095</td>
      <td>2.4821</td>
      <td>10.9149</td>
      <td>6.7593</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>16.7238</td>
      <td>3.1704</td>
      <td>9.1443</td>
      <td>2.4821</td>
      <td>4.3206</td>
      <td>3.6533</td>
      <td>4.8856</td>
      <td>10.5574</td>
      <td>6.0253</td>
      <td>7.8940</td>
      <td>...</td>
      <td>4.6727</td>
      <td>6.9805</td>
      <td>5.4212</td>
      <td>3.6533</td>
      <td>3.4468</td>
      <td>3.6533</td>
      <td>8.9874</td>
      <td>3.4468</td>
      <td>9.7828</td>
      <td>3.4468</td>
    </tr>
    <tr>
      <th>TCGA-86-8076</th>
      <td>15.6783</td>
      <td>4.7422</td>
      <td>9.5150</td>
      <td>3.2271</td>
      <td>7.2321</td>
      <td>5.1146</td>
      <td>7.2174</td>
      <td>10.3567</td>
      <td>7.2467</td>
      <td>10.4765</td>
      <td>...</td>
      <td>4.8254</td>
      <td>7.9800</td>
      <td>3.6417</td>
      <td>3.9275</td>
      <td>3.6417</td>
      <td>5.0478</td>
      <td>9.5810</td>
      <td>3.0118</td>
      <td>8.6640</td>
      <td>4.7844</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-86-8073</th>
      <td>16.3773</td>
      <td>4.0817</td>
      <td>9.4234</td>
      <td>6.6519</td>
      <td>6.7641</td>
      <td>4.2611</td>
      <td>7.1467</td>
      <td>11.3497</td>
      <td>7.0724</td>
      <td>8.7741</td>
      <td>...</td>
      <td>8.6449</td>
      <td>9.2445</td>
      <td>5.9388</td>
      <td>6.8492</td>
      <td>5.8132</td>
      <td>7.2074</td>
      <td>10.2410</td>
      <td>5.1126</td>
      <td>10.1390</td>
      <td>8.2478</td>
    </tr>
    <tr>
      <th>TCGA-MN-A4N4</th>
      <td>14.9411</td>
      <td>6.7210</td>
      <td>11.9107</td>
      <td>8.8283</td>
      <td>3.4947</td>
      <td>3.0746</td>
      <td>4.1736</td>
      <td>9.6396</td>
      <td>6.6811</td>
      <td>9.3538</td>
      <td>...</td>
      <td>10.1845</td>
      <td>7.5083</td>
      <td>5.0939</td>
      <td>5.7178</td>
      <td>6.2061</td>
      <td>6.9167</td>
      <td>9.4672</td>
      <td>3.3143</td>
      <td>10.0583</td>
      <td>8.4566</td>
    </tr>
    <tr>
      <th>TCGA-53-7626</th>
      <td>16.7343</td>
      <td>5.2575</td>
      <td>9.5379</td>
      <td>2.4821</td>
      <td>5.7120</td>
      <td>5.0910</td>
      <td>5.6287</td>
      <td>10.7802</td>
      <td>9.3334</td>
      <td>9.3085</td>
      <td>...</td>
      <td>5.4842</td>
      <td>7.9278</td>
      <td>3.8286</td>
      <td>3.7563</td>
      <td>3.8286</td>
      <td>6.0927</td>
      <td>10.0507</td>
      <td>4.2816</td>
      <td>9.4627</td>
      <td>7.4014</td>
    </tr>
    <tr>
      <th>TCGA-44-A47G</th>
      <td>15.6622</td>
      <td>3.7345</td>
      <td>9.8660</td>
      <td>4.1045</td>
      <td>6.2214</td>
      <td>4.9759</td>
      <td>4.2980</td>
      <td>10.3975</td>
      <td>7.8841</td>
      <td>10.1369</td>
      <td>...</td>
      <td>4.9218</td>
      <td>7.7402</td>
      <td>5.9326</td>
      <td>4.7450</td>
      <td>3.9944</td>
      <td>5.3854</td>
      <td>7.4843</td>
      <td>4.4650</td>
      <td>9.5817</td>
      <td>4.6804</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>14.8136</td>
      <td>4.3058</td>
      <td>9.7702</td>
      <td>6.8425</td>
      <td>5.2016</td>
      <td>3.9293</td>
      <td>7.1699</td>
      <td>8.8889</td>
      <td>6.1528</td>
      <td>7.6565</td>
      <td>...</td>
      <td>6.4289</td>
      <td>7.9499</td>
      <td>4.0012</td>
      <td>6.4059</td>
      <td>3.7694</td>
      <td>3.6792</td>
      <td>8.5268</td>
      <td>2.4821</td>
      <td>10.8228</td>
      <td>8.6068</td>
    </tr>
  </tbody>
</table>
<p>498 rows Ã— 5338 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># df_renamed.to_csv(os.path.join(raw_data_dir,&quot;expression_data_filtered.csv&quot;), index=True)</span>
</pre></div>
</div>
</div>
</div>
<p>Letâ€™s look at the distribution again, for the filtered data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the distribution of gene expression levels </span>

<span class="c1"># Calculate the mean expression level for each gene</span>
<span class="n">gene_means</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Plot the distribution of gene expression levels</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">gene_means</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mean Gene Expression&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Gene Expression Levels&#39;</span><span class="p">)</span>

<span class="c1"># mean of the gene means</span>
<span class="n">threshold</span> <span class="o">=</span> <span class="n">gene_means</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Mean = </span><span class="si">{</span><span class="n">threshold</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/5ce692f91c47186d588dedebec04b8287cff87aaf0cc73eb44c57e377c6d4a6f.png" src="../_images/5ce692f91c47186d588dedebec04b8287cff87aaf0cc73eb44c57e377c6d4a6f.png" />
</div>
</div>
<p>Letâ€™s see the most extreme cases to get a better idea what is in the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sort and get the top 10 and bottom 10 columns based on their mean values</span>
<span class="n">top_10_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="n">bottom_10_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># Combine top 10 and bottom 10 columns into one DataFrame</span>
<span class="n">combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">top_10_columns</span><span class="p">,</span> <span class="n">bottom_10_columns</span><span class="p">])</span>

<span class="c1"># Plot the results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">combined</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;bar&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;skyblue&#39;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">10</span> <span class="k">else</span> <span class="s1">&#39;lightcoral&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">20</span><span class="p">)])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Top 10 and Bottom 10 Genes Based on Mean Expression Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Mean Value&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a1bd13c4db0d9fa2c9b2e4b9a609af3d7408f3e7921effd95821624aa0e8811e.png" src="../_images/a1bd13c4db0d9fa2c9b2e4b9a609af3d7408f3e7921effd95821624aa0e8811e.png" />
</div>
</div>
<section id="further-data-investigation">
<h2>Further Data Investigation<a class="headerlink" href="#further-data-investigation" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select numerical columns</span>
<span class="n">numerical_columns</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;float64&#39;</span><span class="p">,</span> <span class="s1">&#39;int64&#39;</span><span class="p">])</span>

<span class="c1"># Calculate the mean of each column for the numerical columns only</span>
<span class="n">means</span> <span class="o">=</span> <span class="n">numerical_columns</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

<span class="c1"># Get the top 8 and bottom 8 columns based on mean values</span>
<span class="n">top_8_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">bottom_8_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">nsmallest</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># Combine top 8 and bottom 8 columns for plotting</span>
<span class="n">columns_to_plot</span> <span class="o">=</span> <span class="n">top_8_columns</span> <span class="o">+</span> <span class="n">bottom_8_columns</span>

<span class="c1"># Filter the numerical columns to only include those to plot</span>
<span class="n">filtered_numerical_columns</span> <span class="o">=</span> <span class="n">numerical_columns</span><span class="p">[</span><span class="n">columns_to_plot</span><span class="p">]</span>

<span class="c1"># Calculate the number of rows needed for subplots based on the number of selected columns</span>
<span class="n">num_plots</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_numerical_columns</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="n">num_rows</span> <span class="o">=</span> <span class="p">(</span><span class="n">num_plots</span> <span class="o">//</span> <span class="mi">4</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">num_plots</span> <span class="o">%</span> <span class="mi">4</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Ensure there is an extra row if there are leftovers</span>

<span class="c1"># Plot histograms for each selected numerical column</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">num_rows</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">num_rows</span><span class="p">))</span>  <span class="c1"># Adjust width and height as needed</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Distribution of Top 8 &amp; Bottom 8 Genes Mean Expression&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">filtered_numerical_columns</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">filtered_numerical_columns</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;skyblue&#39;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">8</span> <span class="k">else</span> <span class="s1">&#39;lightcoral&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
    
    <span class="c1"># Adding mean and median lines</span>
    <span class="n">mean_val</span> <span class="o">=</span> <span class="n">filtered_numerical_columns</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">median_val</span> <span class="o">=</span> <span class="n">filtered_numerical_columns</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">median</span><span class="p">()</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">mean_val</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">median_val</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">({</span><span class="s1">&#39;Mean&#39;</span><span class="p">:</span> <span class="n">mean_val</span><span class="p">,</span> <span class="s1">&#39;Median&#39;</span><span class="p">:</span> <span class="n">median_val</span><span class="p">})</span>

<span class="c1"># Hide any unused axes if the number of plots isn&#39;t a perfect multiple of 4</span>
<span class="k">if</span> <span class="n">num_plots</span> <span class="o">%</span> <span class="mi">4</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">num_plots</span><span class="p">:]:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">(</span><span class="n">rect</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">])</span>  <span class="c1"># Adjust layout to make room for the suptitle</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/4d66d50372835c46e591dabe80a2572d2dde0fc023385d916c06bfce351fa159.png" src="../_images/4d66d50372835c46e591dabe80a2572d2dde0fc023385d916c06bfce351fa159.png" />
</div>
</div>
</section>
</section>
<section id="correlation-matrix-calculation">
<h1>4. Correlation Matrix Calculation<a class="headerlink" href="#correlation-matrix-calculation" title="Permalink to this heading">#</a></h1>
<p>In order to create a biological network, we have to calculate the correlation matrix of the gene expression data.<br />
The correlation matrix shows how strongly pairs of genes are related. <br />
The assumption is that if genes are correlated, then this means they influence each others behaviour.</p>
<p>Using the correlation matrix, we can build a gene expression network. In this network, genes are connected based on significant correlations.</p>
<ul class="simple">
<li><p>nodes: genes</p></li>
<li><p>edges: highly correlated genes (above a given threshold)</p></li>
<li><p>edge-weights: correlation values</p></li>
</ul>
<p>There are a few correlation metrics one could consider:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient">Pearson</a></p>
<ul>
<li><p>O(n^2) complexity, fast for large datasets</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient">Spearman</a></p>
<ul>
<li><p>O(n^2 log n) complexity, relatively fast but can be slower than Pearson</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Biweight_midcorrelation">Absolute biweight midcorrelation</a></p>
<ul>
<li><p>Robust but slower than Pearson and Spearman, suitable for datasets with outliers</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># function to calculate absolute biweight midcorrelation </span>
<span class="k">def</span> <span class="nf">abs_bicorr</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="c1"># Select only numeric data</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">_get_numeric_data</span><span class="p">()</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">columns</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">cols</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">na_value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">T</span>

    <span class="n">K</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span>
    <span class="n">correl</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">K</span><span class="p">,</span> <span class="n">K</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

    <span class="c1"># Calculate biweight midcovariance</span>
    <span class="n">bicorr</span> <span class="o">=</span> <span class="n">astropy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">biweight_midcovariance</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">modify_sample_size</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">K</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">K</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">j</span><span class="p">:</span>
                <span class="n">correl</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">denominator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">bicorr</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">bicorr</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">j</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">denominator</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">correl</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">bicorr</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">/</span> <span class="n">denominator</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">correl</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Or handle it in another appropriate way</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">correl</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">cols</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Consider loading in precomputed correlation matrices and skipping the cell after.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the entire dictionary of correlation matrices from a pickle file</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">intermediate_data_dir</span><span class="p">,</span><span class="s2">&quot;correlation_matrices.pkl&quot;</span><span class="p">),</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">correlation_matrices</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

<span class="c1"># Verify the loaded data</span>
<span class="nb">print</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;pearson&#39;, &#39;spearman&#39;, &#39;biweight_midcorrelation&#39;])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># # Dictionary to store different correlation matrices</span>
<span class="c1"># correlation_matrices = {}</span>

<span class="c1"># # Pearson correlation - O(n^2) complexity, fast for large datasets</span>
<span class="c1"># correlation_matrices[&#39;pearson&#39;] = df_renamed.corr(method=&#39;pearson&#39;)</span>

<span class="c1"># # Spearman rank correlation -  O(n^2 log n) complexity, relatively fast but can be slower than Pearson</span>
<span class="c1"># correlation_matrices[&#39;spearman&#39;] = df_renamed.corr(method=&#39;spearman&#39;)</span>

<span class="c1"># # Biweight midcorrelation -  Robust but slower than Pearson and Spearman, suitable for datasets with outliers</span>
<span class="c1"># correlation_matrices[&#39;biweight_midcorrelation&#39;] = abs_bicorr(df_renamed)</span>

<span class="c1"># # Print the keys of the correlation matrices to verify</span>
<span class="c1"># print(&quot;Correlation matrices calculated:&quot;)</span>
<span class="c1"># print(correlation_matrices.keys())</span>


<span class="c1"># # Save the entire dictionary of correlation matrices as a pickle file</span>
<span class="c1"># with open(os.path.join(intermediate_data_dir,&quot;correlation_matrices.pkl&quot;), &#39;wb&#39;) as f:</span>
<span class="c1">#     pickle.dump(correlation_matrices, f)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.cluster.hierarchy</span> <span class="kn">import</span> <span class="n">linkage</span><span class="p">,</span> <span class="n">leaves_list</span>

<span class="k">def</span> <span class="nf">plot_correlation_matrices</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">):</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>  <span class="c1"># Adjust to 1 row and 3 columns</span>
    <span class="n">axes</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">matrix</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="c1"># Perform hierarchical clustering</span>
        <span class="n">Z</span> <span class="o">=</span> <span class="n">linkage</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">)</span>  <span class="c1"># You can use other methods like &#39;single&#39;, &#39;complete&#39;, etc.</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">leaves_list</span><span class="p">(</span><span class="n">Z</span><span class="p">)</span>
        
        <span class="c1"># Reorder matrix</span>
        <span class="n">ordered_matrix</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">idx</span><span class="p">]</span>
        
        <span class="c1"># Plot heatmap</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">ordered_matrix</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">xticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">key</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1"> Correlation Matrix&#39;</span><span class="p">)</span>
        
        <span class="c1"># Set square aspect ratio</span>
        <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
    
    <span class="c1"># Hide any unused subplots</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">axes</span><span class="p">)):</span>
        <span class="n">axes</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We are not going to run this script, because it take a while.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># A correlation_matrices is a dictionary of correlation matrices</span>
<span class="c1"># plot_correlation_matrices(correlation_matrices)</span>
</pre></div>
</div>
</div>
</div>
<p><img alt="Correlation Matrices" src="../_images/correlation_matrices_figure.png" /></p>
</section>
<section id="graph-construction-from-correlation-matrices">
<h1>5. Graph Construction from Correlation Matrices<a class="headerlink" href="#graph-construction-from-correlation-matrices" title="Permalink to this heading">#</a></h1>
<p>In this stage we are leaving behind the raw data, and we are going to start to work with network objects.<br />
We are using the <a class="reference external" href="https://networkx.org/documentation/stable/reference/index.html">NetworkX</a> library for this.</p>
<p>We are going to run a function <code class="docutils literal notranslate"><span class="pre">create_graph_from_correlation</span></code>.</p>
<p>The function starts by creating an empty graph G. Then iterates through the columns of the correlation matrix and adds each column name as a node in the graph. This means each gene (or feature) in your dataset becomes a node in the graph.</p>
<p>The function iterates over the upper triangle of the correlation matrix (excluding the diagonal) to avoid redundancy and self-loops.</p>
<p>For each pair of nodes (i, j), it checks if the absolute value of the correlation coefficient between them is greater than or equal to the specified threshold.
If the condition is met, an edge is added between the nodes i and j with the correlation coefficient as the weight of the edge. This signifies a strong correlation (positive or negative) between the two nodes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_graph_from_correlation</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a graph from a correlation matrix using a specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    correlation_matrix (pd.DataFrame): DataFrame containing the correlation matrix.</span>
<span class="sd">    threshold (float): Threshold for including edges based on correlation value.</span>

<span class="sd">    Returns:</span>
<span class="sd">    G (nx.Graph): Graph created from the correlation matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>

    <span class="c1"># Add nodes</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">G</span><span class="o">.</span><span class="n">add_node</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

    <span class="c1"># Add edges with weights above the threshold</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>  <span class="c1"># Ignore the diagonal elements</span>
                <span class="n">weight</span> <span class="o">=</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">weight</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">:</span>
                    <span class="n">G</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">weight</span><span class="o">=</span><span class="n">weight</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## correlation_matrices is a dictionary containing correlation matrices</span>
<span class="c1"># pearson_graph = create_graph_from_correlation(correlation_matrices[&#39;pearson&#39;], threshold=0.8)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># nx.write_gml(pearson_graph, os.path.join(data_dir,&#39;gene_coexpression_network_pearson.gml&#39;))</span>
<span class="n">pearson_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">read_gml</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">intermediate_data_dir</span><span class="p">,</span><span class="s1">&#39;gene_coexpression_network_pearson.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>Now letâ€™s go through a few useful NetworkX functions and create a <code class="docutils literal notranslate"><span class="pre">print_graph_info()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">print_graph_info</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prints basic information about the graph.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of nodes: </span><span class="si">{</span><span class="n">G</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of edges: </span><span class="si">{</span><span class="n">G</span><span class="o">.</span><span class="n">number_of_edges</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample nodes:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">)[:</span><span class="mi">10</span><span class="p">])</span>  <span class="c1"># Print first 10 nodes as a sample</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample edges:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))[:</span><span class="mi">10</span><span class="p">])</span>  <span class="c1"># Print first 10 edges as a sample</span>
    
    <span class="n">info_str</span> <span class="o">=</span> <span class="s2">&quot;Graph type: &quot;</span>
    <span class="n">is_directed</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">is_directed</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">is_directed</span><span class="p">:</span>
        <span class="n">info_str</span> <span class="o">+=</span> <span class="s2">&quot;directed&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">info_str</span> <span class="o">+=</span> <span class="s2">&quot;undirected&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">info_str</span><span class="p">)</span>

    <span class="c1"># Check for self-loops</span>
    <span class="n">self_loops</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">selfloop_edges</span><span class="p">(</span><span class="n">G</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">self_loops</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of self-loops: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">self_loops</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Self-loops:&quot;</span><span class="p">,</span> <span class="n">self_loops</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No self-loops in the graph.&quot;</span><span class="p">)</span>

    <span class="c1"># density of the graph</span>
    <span class="n">density</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">density</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Graph density: </span><span class="si">{</span><span class="n">density</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Find and print the number of connected components</span>
    <span class="n">num_connected_components</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">number_connected_components</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of connected components: </span><span class="si">{</span><span class="n">num_connected_components</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Calculate and print the clustering coefficient of the graph</span>
    <span class="n">clustering_coeff</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">average_clustering</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Average clustering coefficient: </span><span class="si">{</span><span class="n">clustering_coeff</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 5338
Number of edges: 13449
Sample nodes: [&#39;A2M&#39;, &#39;A2ML1&#39;, &#39;A4GALT&#39;, &#39;AACSP1&#39;, &#39;AADAC&#39;, &#39;AADACP1&#39;, &#39;AARD&#39;, &#39;AASS&#39;, &#39;AATBC&#39;, &#39;AATK&#39;]
Sample edges: [(&#39;A2M&#39;, &#39;ABCA8&#39;, {&#39;weight&#39;: 0.8308274641648915}), (&#39;A2M&#39;, &#39;ADAMTS8&#39;, {&#39;weight&#39;: 0.8036704002698273}), (&#39;A2M&#39;, &#39;AOC3&#39;, {&#39;weight&#39;: 0.8767411078325781}), (&#39;A2M&#39;, &#39;C7&#39;, {&#39;weight&#39;: 0.8013688242083488}), (&#39;A2M&#39;, &#39;INMT&#39;, {&#39;weight&#39;: 0.8117391230502314}), (&#39;A2M&#39;, &#39;ITGA8&#39;, {&#39;weight&#39;: 0.8724141478160804}), (&#39;A2M&#39;, &#39;MFAP4&#39;, {&#39;weight&#39;: 0.8444697225886688}), (&#39;A2M&#39;, &#39;PKNOX2&#39;, {&#39;weight&#39;: 0.8062841461653377}), (&#39;A2M&#39;, &#39;SCN7A&#39;, {&#39;weight&#39;: 0.8382334505662205}), (&#39;A2M&#39;, &#39;SLIT2&#39;, {&#39;weight&#39;: 0.8041515706739126})]
Graph type: undirected
No self-loops in the graph.
Graph density: 0.0009441569992192751
Number of connected components: 4392
Average clustering coefficient: 0.1045798392696385
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># biweight_graph = create_graph_from_correlation(correlation_matrices[&#39;biweight_midcorrelation&#39;], threshold=0.8)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to visualize the graph</span>
<span class="k">def</span> <span class="nf">visualize_graph</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Gene Co-expression Network&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes the graph using Matplotlib and NetworkX.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): Graph to visualize.</span>
<span class="sd">    title (str): Title of the plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># k controls the distance between nodes</span>
    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_nodes</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">node_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">node_color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_edges</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># # Visualize the graph</span>
<span class="c1"># visualize_graph(pearson_graph, title=&#39;Pearson Correlation Network (Threshold = 0.8)&#39;)</span>
</pre></div>
</div>
</div>
</div>
<p><img alt="Full Graph" src="../_images/full_gene_coexpression_network.png" /></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># visualize_graph(biweight_graph, title=&#39;Biweight MidCorrelation Network (Threshold = 0.8)&#39;)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="network-cleaning">
<h1>6. Network Cleaning<a class="headerlink" href="#network-cleaning" title="Permalink to this heading">#</a></h1>
<p>Now that we have created the gene co-expression networks, we can further clean and analyse the network.</p>
<p>We can still use functions of NetworkX specifically made for this.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">clean_graph</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Cleans the graph by performing several cleaning steps:</span>
<span class="sd">    - Removes unconnected nodes (isolates)</span>
<span class="sd">    - Removes self-loops</span>
<span class="sd">    - Removes nodes with a degree below a specified threshold</span>
<span class="sd">    - Keeps only the largest connected component (optional)</span>

<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph to clean.</span>
<span class="sd">    degree_threshold (int): Minimum degree for nodes to keep.</span>
<span class="sd">    keep_largest_component (bool): Whether to keep only the largest connected component.</span>

<span class="sd">    Returns:</span>
<span class="sd">    G (nx.Graph): Cleaned graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>  <span class="c1"># Work on a copy of the graph to avoid modifying the original graph</span>

    <span class="c1"># Remove self-loops</span>
    <span class="n">G</span><span class="o">.</span><span class="n">remove_edges_from</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">selfloop_edges</span><span class="p">(</span><span class="n">G</span><span class="p">))</span>

    <span class="c1"># Remove nodes with no edges (isolates)</span>
    <span class="n">G</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">isolates</span><span class="p">(</span><span class="n">G</span><span class="p">)))</span>

    <span class="c1"># Remove nodes with degree below the threshold</span>
    <span class="n">low_degree_nodes</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="nb">dict</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">degree</span> <span class="o">&lt;</span> <span class="n">degree_threshold</span><span class="p">]</span>
    <span class="n">G</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="n">low_degree_nodes</span><span class="p">)</span>

    <span class="c1"># Keep only the largest connected component</span>
    <span class="k">if</span> <span class="n">keep_largest_component</span><span class="p">:</span>
        <span class="n">largest_cc</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">connected_components</span><span class="p">(</span><span class="n">G</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="nb">len</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">subgraph</span><span class="p">(</span><span class="n">largest_cc</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">pearson_graph_cleaned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">pearson_graph_cleaned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network - Cleaned&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1c99961626f69da6f4cb497eb6141e464fe7ca35dcc71ee724fcde9484198d71.png" src="../_images/1c99961626f69da6f4cb497eb6141e464fe7ca35dcc71ee724fcde9484198d71.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph_cleaned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 1114
Number of edges: 13449
Sample nodes: [&#39;A2M&#39;, &#39;ABCA8&#39;, &#39;ABCC9&#39;, &#39;ACKR1&#39;, &#39;ACKR2&#39;, &#39;ACOXL&#39;, &#39;ACOXL-AS1&#39;, &#39;ACTG2&#39;, &#39;ADAM12&#39;, &#39;ADAMTS12&#39;]
Sample edges: [(&#39;A2M&#39;, &#39;ABCA8&#39;, {&#39;weight&#39;: 0.8308274641648915}), (&#39;A2M&#39;, &#39;ADAMTS8&#39;, {&#39;weight&#39;: 0.8036704002698273}), (&#39;A2M&#39;, &#39;AOC3&#39;, {&#39;weight&#39;: 0.8767411078325781}), (&#39;A2M&#39;, &#39;C7&#39;, {&#39;weight&#39;: 0.8013688242083488}), (&#39;A2M&#39;, &#39;INMT&#39;, {&#39;weight&#39;: 0.8117391230502314}), (&#39;A2M&#39;, &#39;ITGA8&#39;, {&#39;weight&#39;: 0.8724141478160804}), (&#39;A2M&#39;, &#39;MFAP4&#39;, {&#39;weight&#39;: 0.8444697225886688}), (&#39;A2M&#39;, &#39;PKNOX2&#39;, {&#39;weight&#39;: 0.8062841461653377}), (&#39;A2M&#39;, &#39;SCN7A&#39;, {&#39;weight&#39;: 0.8382334505662205}), (&#39;A2M&#39;, &#39;SLIT2&#39;, {&#39;weight&#39;: 0.8041515706739126})]
Graph type: undirected
No self-loops in the graph.
Graph density: 0.021693999912894935
Number of connected components: 168
Average clustering coefficient: 0.5011195529814455
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">pearson_graph_pruned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network - Pruned&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/5a9dc17cee9199f1a3eed022f308373ba0dbaf6a65d4a1a5e9fb81c980e5c4ee.png" src="../_images/5a9dc17cee9199f1a3eed022f308373ba0dbaf6a65d4a1a5e9fb81c980e5c4ee.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># save the graph as a GML file for future use</span>
<span class="c1">#nx.write_gml(pearson_graph_pruned, os.path.join(intermediate_data_dir,&quot;gene_coexpression_network.gml&quot;))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 305
Number of edges: 4033
Sample nodes: [&#39;SPN&#39;, &#39;SCML4&#39;, &#39;IL2RG&#39;, &#39;CCL23&#39;, &#39;TLR7&#39;, &#39;FAM30A&#39;, &#39;LILRB4&#39;, &#39;IGKV4-1&#39;, &#39;CXCR6&#39;, &#39;FCGR1A&#39;]
Sample edges: [(&#39;SPN&#39;, &#39;CD300LF&#39;, {&#39;weight&#39;: 0.8054549326650037}), (&#39;SPN&#39;, &#39;CYBB&#39;, {&#39;weight&#39;: 0.8225847332114441}), (&#39;SPN&#39;, &#39;DOK2&#39;, {&#39;weight&#39;: 0.8694303780879173}), (&#39;SPN&#39;, &#39;FGR&#39;, {&#39;weight&#39;: 0.8239102785759777}), (&#39;SPN&#39;, &#39;HCK&#39;, {&#39;weight&#39;: 0.8015665561321501}), (&#39;SPN&#39;, &#39;JAML&#39;, {&#39;weight&#39;: 0.9034237904179987}), (&#39;SPN&#39;, &#39;MAP3K5-AS2&#39;, {&#39;weight&#39;: 0.8098711495923998}), (&#39;SPN&#39;, &#39;MNDA&#39;, {&#39;weight&#39;: 0.8269221117933395}), (&#39;SPN&#39;, &#39;P2RY13&#39;, {&#39;weight&#39;: 0.823261045280138}), (&#39;SPN&#39;, &#39;PRAM1&#39;, {&#39;weight&#39;: 0.8016230886119241})]
Graph type: undirected
No self-loops in the graph.
Graph density: 0.08699309749784297
Number of connected components: 1
Average clustering coefficient: 0.6766781614263463
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_degree_distribution</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the degree distribution of the graph using Seaborn.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">degrees</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">()]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">degrees</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Degree Distribution&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Degree&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
<span class="n">plot_degree_distribution</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 305
Number of edges: 4033
Sample nodes: [&#39;SPN&#39;, &#39;SCML4&#39;, &#39;IL2RG&#39;, &#39;CCL23&#39;, &#39;TLR7&#39;, &#39;FAM30A&#39;, &#39;LILRB4&#39;, &#39;IGKV4-1&#39;, &#39;CXCR6&#39;, &#39;FCGR1A&#39;]
Sample edges: [(&#39;SPN&#39;, &#39;CD300LF&#39;, {&#39;weight&#39;: 0.8054549326650037}), (&#39;SPN&#39;, &#39;CYBB&#39;, {&#39;weight&#39;: 0.8225847332114441}), (&#39;SPN&#39;, &#39;DOK2&#39;, {&#39;weight&#39;: 0.8694303780879173}), (&#39;SPN&#39;, &#39;FGR&#39;, {&#39;weight&#39;: 0.8239102785759777}), (&#39;SPN&#39;, &#39;HCK&#39;, {&#39;weight&#39;: 0.8015665561321501}), (&#39;SPN&#39;, &#39;JAML&#39;, {&#39;weight&#39;: 0.9034237904179987}), (&#39;SPN&#39;, &#39;MAP3K5-AS2&#39;, {&#39;weight&#39;: 0.8098711495923998}), (&#39;SPN&#39;, &#39;MNDA&#39;, {&#39;weight&#39;: 0.8269221117933395}), (&#39;SPN&#39;, &#39;P2RY13&#39;, {&#39;weight&#39;: 0.823261045280138}), (&#39;SPN&#39;, &#39;PRAM1&#39;, {&#39;weight&#39;: 0.8016230886119241})]
Graph type: undirected
No self-loops in the graph.
Graph density: 0.08699309749784297
Number of connected components: 1
Average clustering coefficient: 0.6766781614263463
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/98ed257f4b265bc811ece72b015764ef4b93ad9fb1e444ba65c6efce978ee9cd.png" src="../_images/98ed257f4b265bc811ece72b015764ef4b93ad9fb1e444ba65c6efce978ee9cd.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">visualize_edge_weight_distribution</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes the distribution of edge weights using a histogram and KDE plot.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    edge_weights (list): List of edge weights.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">edge_weights</span> <span class="o">=</span> <span class="p">[</span><span class="n">G</span><span class="p">[</span><span class="n">u</span><span class="p">][</span><span class="n">v</span><span class="p">][</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">()]</span>
    <span class="c1"># Histogram</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">edge_weights</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Edge Weights&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Edge Weight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the distribution of edge weights</span>
<span class="n">visualize_edge_weight_distribution</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/37d213fe14fb8fcf9e29f9aebcd375cfb81490304954b2089277729fa589d497.png" src="../_images/37d213fe14fb8fcf9e29f9aebcd375cfb81490304954b2089277729fa589d497.png" />
</div>
</div>
</section>
<section id="sparsification-methods">
<h1>7. Sparsification methods<a class="headerlink" href="#sparsification-methods" title="Permalink to this heading">#</a></h1>
<p>With sparsification we aim to reduce the number of edges in a network while preserving important structural properties.</p>
<ul class="simple">
<li><p>Edge Sampling: Randomly removes a fraction of edges.</p></li>
<li><p>Thresholding: Removes edges with weights below a certain threshold.</p></li>
<li><p>Degree-based Sparsification</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">threshold_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by removing edges below the specified weight threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    threshold (float): The weight threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">graph_copy</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">graph_copy</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">((</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">graph_copy</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">d</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;weight&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sparsified_graph</span>

<span class="k">def</span> <span class="nf">top_percentage_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">top_percentage</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by keeping the top percentage of edges by weight.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    top_percentage (float): The percentage of top-weight edges to keep.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">graph_copy</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sorted_edges</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">graph_copy</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;weight&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">top_edges_count</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sorted_edges</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">top_percentage</span> <span class="o">/</span> <span class="mi">100</span><span class="p">)))</span>
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">graph_copy</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">sorted_edges</span><span class="p">[:</span><span class="n">top_edges_count</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">sparsified_graph</span>


<span class="k">def</span> <span class="nf">remove_by_degree</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">min_degree</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by removing nodes with degree below the specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    min_degree (int): The minimum degree threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">graph_copy</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">nodes_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="nb">dict</span><span class="p">(</span><span class="n">graph_copy</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">degree</span> <span class="o">&lt;</span> <span class="n">min_degree</span><span class="p">]</span>
    
    <span class="n">graph_copy</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="n">nodes_to_remove</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">graph_copy</span>

<span class="k">def</span> <span class="nf">knn_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">k</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by keeping only the top-k edges with the highest weights for each node.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    k (int): The number of nearest neighbors to keep for each node.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">graph_copy</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">graph_copy</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph_copy</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
        <span class="n">edges</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">graph_copy</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;weight&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">edges</span><span class="p">[:</span><span class="n">k</span><span class="p">])</span>
    
    <span class="k">return</span> <span class="n">sparsified_graph</span>


<span class="k">def</span> <span class="nf">spanning_tree_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by creating a minimum spanning tree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">graph_copy</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">nx</span><span class="o">.</span><span class="n">minimum_spanning_tree</span><span class="p">(</span><span class="n">graph_copy</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s1">&#39;weight&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">analyze_and_plot_density</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates and plots the density of the graph for a predefined series of thresholds.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>

<span class="sd">    Returns:</span>
<span class="sd">    densities (list of float): Densities of the graph at each threshold.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.7</span> <span class="o">+</span> <span class="n">i</span> <span class="o">*</span> <span class="mf">0.01</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">31</span><span class="p">)]</span>
    <span class="n">densities</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
        <span class="n">filtered_edges</span> <span class="o">=</span> <span class="p">[(</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">]</span>
        <span class="n">temp_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
        <span class="n">temp_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">filtered_edges</span><span class="p">)</span>
        <span class="n">densities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">density</span><span class="p">(</span><span class="n">temp_graph</span><span class="p">))</span>

    <span class="c1"># Plot the densities</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">thresholds</span><span class="p">,</span> <span class="n">densities</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Threshold&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Density vs. Threshold&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">densities</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">densities</span> <span class="o">=</span> <span class="n">analyze_and_plot_density</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b637e9e19ecaf06f129d576eaa8e69e8ed2f5bfad6413b2c8e6d77e7e787d97b.png" src="../_images/b637e9e19ecaf06f129d576eaa8e69e8ed2f5bfad6413b2c8e6d77e7e787d97b.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize a dictionary to store graphs</span>
<span class="n">graphs</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pearson_graph_pruned</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>  <span class="c1"># Store the original graph</span>

<span class="c1"># Apply sparsification methods</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;threshold&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">threshold_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.82</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;top_10_percent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">top_percentage_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">top_percentage</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;degree_below_3&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">remove_by_degree</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">min_degree</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;knn_5&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;spanning_tree&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spanning_tree_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">])</span>


<span class="c1"># Example visualization</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="s1">&#39;Original Graph&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;threshold&#39;</span><span class="p">],</span> <span class="s1">&#39;Thresholded Graph (weight &gt; 0.82)&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;top_10_percent&#39;</span><span class="p">],</span> <span class="s1">&#39;Top 10</span><span class="si">% E</span><span class="s1">dges by Weight&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;degree_below_3&#39;</span><span class="p">],</span> <span class="s1">&#39;Degree Below 3&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;knn_5&#39;</span><span class="p">],</span> <span class="s1">&#39;K-Nearest Neighbors (k=5)&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;spanning_tree&#39;</span><span class="p">],</span> <span class="s1">&#39;Minimum Spanning Tree&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/617c10c77cfc2358dbfd11762d3123f7c76f3b30b73b57b3f9a40ee9afdffc76.png" src="../_images/617c10c77cfc2358dbfd11762d3123f7c76f3b30b73b57b3f9a40ee9afdffc76.png" />
<img alt="../_images/0201c23ca68f02f6c283b3efc193937ae6a3aa96cd92bfb9b76d5a036eb00440.png" src="../_images/0201c23ca68f02f6c283b3efc193937ae6a3aa96cd92bfb9b76d5a036eb00440.png" />
<img alt="../_images/63adf56e1830e9703b16309c4c76fd04d1ba055e9ad4a543f84f6483c590c910.png" src="../_images/63adf56e1830e9703b16309c4c76fd04d1ba055e9ad4a543f84f6483c590c910.png" />
<img alt="../_images/fd470c33dbe69ccd4b26775d27711a7d0a185e8052912dfca7af0053a03fa076.png" src="../_images/fd470c33dbe69ccd4b26775d27711a7d0a185e8052912dfca7af0053a03fa076.png" />
<img alt="../_images/7d4cc020e8dfc7f4214f2344cd99b56a46ddb967e04ef003ddd30559cc3ea39a.png" src="../_images/7d4cc020e8dfc7f4214f2344cd99b56a46ddb967e04ef003ddd30559cc3ea39a.png" />
<img alt="../_images/d75b736d6eb54a249ed5b1702fd9c32ab8333a0691b50a27ad89b72ff271756e.png" src="../_images/d75b736d6eb54a249ed5b1702fd9c32ab8333a0691b50a27ad89b72ff271756e.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;knn_5&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 305
Number of edges: 1078
Sample nodes: [&#39;SPN&#39;, &#39;SCML4&#39;, &#39;IL2RG&#39;, &#39;CCL23&#39;, &#39;TLR7&#39;, &#39;FAM30A&#39;, &#39;LILRB4&#39;, &#39;IGKV4-1&#39;, &#39;CXCR6&#39;, &#39;FCGR1A&#39;]
Sample edges: [(&#39;SPN&#39;, &#39;JAML&#39;, {&#39;weight&#39;: 0.9034237904179987}), (&#39;SPN&#39;, &#39;DOK2&#39;, {&#39;weight&#39;: 0.8694303780879173}), (&#39;SPN&#39;, &#39;SCIMP&#39;, {&#39;weight&#39;: 0.8624303546297505}), (&#39;SPN&#39;, &#39;MNDA&#39;, {&#39;weight&#39;: 0.8269221117933395}), (&#39;SPN&#39;, &#39;FGR&#39;, {&#39;weight&#39;: 0.8239102785759777}), (&#39;SCML4&#39;, &#39;ZNF831&#39;, {&#39;weight&#39;: 0.883224320882079}), (&#39;SCML4&#39;, &#39;ITK&#39;, {&#39;weight&#39;: 0.8813986345201285}), (&#39;SCML4&#39;, &#39;SLAMF6&#39;, {&#39;weight&#39;: 0.8469829399349706}), (&#39;SCML4&#39;, &#39;CD3E&#39;, {&#39;weight&#39;: 0.8438052940264802}), (&#39;SCML4&#39;, &#39;GPR174&#39;, {&#39;weight&#39;: 0.8432111610713862})]
Graph type: undirected
No self-loops in the graph.
Graph density: 0.023252804141501295
Number of connected components: 2
Average clustering coefficient: 0.5245715856842815
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">analyse_knn_effect</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">k_values</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyses the effect of different k values on the network properties.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    k_values (list): List of k values to use for sparsification.</span>

<span class="sd">    Returns:</span>
<span class="sd">    pd.DataFrame: DataFrame containing the analysis results.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;k&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;num_edges&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;avg_degree&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;avg_clustering&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;num_connected_components&#39;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">k_values</span><span class="p">:</span>
        <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
        <span class="n">num_edges</span> <span class="o">=</span> <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">number_of_edges</span><span class="p">()</span>
        <span class="n">avg_degree</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">/</span> <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">()</span>
        <span class="n">avg_clustering</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">average_clustering</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="p">)</span>
        <span class="n">num_connected_components</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">number_connected_components</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="p">)</span>
        
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;num_edges&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num_edges</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;avg_degree&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_degree</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;avg_clustering&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_clustering</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;num_connected_components&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num_connected_components</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_knn_analysis</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the analysis of the effect of different k values on network properties.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    df (pd.DataFrame): DataFrame containing the analysis results.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;num_edges&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Number of Edges vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Edges&#39;</span><span class="p">)</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg_degree&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Average Degree vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Average Degree&#39;</span><span class="p">)</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg_clustering&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Average Clustering Coefficient vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Average Clustering Coefficient&#39;</span><span class="p">)</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;num_connected_components&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Number of Connected Components vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Connected Components&#39;</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>


<span class="n">k_values</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">))</span>  <span class="c1"># Different k values to analyze</span>
<span class="n">analysis_results</span> <span class="o">=</span> <span class="n">analyse_knn_effect</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">k_values</span><span class="p">)</span>

<span class="c1"># Plot the analysis results</span>
<span class="n">plot_knn_analysis</span><span class="p">(</span><span class="n">analysis_results</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/569ec47d3f9a603a2cf5f89c8d33996a219f0e923460328f93f4292a06ee204a.png" src="../_images/569ec47d3f9a603a2cf5f89c8d33996a219f0e923460328f93f4292a06ee204a.png" />
</div>
</div>
</section>
<section id="describing-highly-connected-nodes">
<h1>8. Describing Highly Connected Nodes<a class="headerlink" href="#describing-highly-connected-nodes" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_highest_degree_nodes</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the nodes with the highest degree in the graph.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The NetworkX graph.</span>
<span class="sd">    top_n (int): The number of top nodes to return.</span>

<span class="sd">    Returns:</span>
<span class="sd">    List of tuples: Each tuple contains a node and its degree.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">degrees</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span>
    <span class="n">sorted_degrees</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">degrees</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sorted_degrees</span><span class="p">[:</span><span class="n">top_n</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">fetch_gene_info</span><span class="p">(</span><span class="n">gene_list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches gene information from MyGene.info.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene_list (list): List of gene symbols or Ensembl IDs.</span>

<span class="sd">    Returns:</span>
<span class="sd">    list: List of dictionaries containing gene information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mg</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">MyGeneInfo</span><span class="p">()</span>
    <span class="n">gene_info</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;symbol,ensembl.gene&#39;</span><span class="p">,</span> 
                             <span class="n">fields</span><span class="o">=</span><span class="s1">&#39;name,symbol,entrezgene,summary,disease,pathway&#39;</span><span class="p">,</span> 
                             <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gene_info</span>

<span class="k">def</span> <span class="nf">print_gene_info_with_degree</span><span class="p">(</span><span class="n">top_genes_with_degrees</span><span class="p">,</span> <span class="n">gene_info</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prints gene information including the degree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    top_genes_with_degrees (list): List of tuples containing gene symbols and their degrees.</span>
<span class="sd">    gene_info (list): List of dictionaries containing gene information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="n">top_genes_with_degrees</span><span class="p">:</span>
        <span class="n">info</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_info</span> <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">),</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">info</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene Symbol: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Degree: </span><span class="si">{</span><span class="n">degree</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene Name: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Entrez ID: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;entrezgene&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Summary: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;summary&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;disease&#39;</span> <span class="ow">in</span> <span class="n">info</span><span class="p">:</span>
                <span class="n">diseases</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">d</span><span class="p">[</span><span class="s1">&#39;term&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="s1">&#39;disease&#39;</span><span class="p">]])</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Diseases: </span><span class="si">{</span><span class="n">diseases</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Diseases: N/A&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;pathway&#39;</span> <span class="ow">in</span> <span class="n">info</span><span class="p">:</span>
                <span class="n">pathways</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;pathway&#39;</span><span class="p">],</span> <span class="nb">dict</span><span class="p">):</span>
                    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="s1">&#39;pathway&#39;</span><span class="p">]:</span>
                        <span class="n">pathway_data</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="s1">&#39;pathway&#39;</span><span class="p">][</span><span class="n">key</span><span class="p">]</span>
                        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                            <span class="n">pathways</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">p</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pathway_data</span> <span class="k">if</span> <span class="s1">&#39;name&#39;</span> <span class="ow">in</span> <span class="n">p</span><span class="p">])</span>
                        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span> <span class="ow">and</span> <span class="s1">&#39;name&#39;</span> <span class="ow">in</span> <span class="n">pathway_data</span><span class="p">:</span>
                            <span class="n">pathways</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">])</span>
                        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                            <span class="n">pathways</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Pathways: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pathways</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">pathways</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Pathways: N/A&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene not found: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Degree: </span><span class="si">{</span><span class="n">degree</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">top_genes_with_degrees</span> <span class="o">=</span> <span class="n">get_highest_degree_nodes</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">gene_symbols</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="n">top_genes_with_degrees</span><span class="p">]</span>

<span class="c1"># Fetch gene information</span>
<span class="n">gene_info</span> <span class="o">=</span> <span class="n">fetch_gene_info</span><span class="p">(</span><span class="n">gene_symbols</span><span class="p">)</span>

<span class="c1"># Print gene information including degree</span>
<span class="n">print_gene_info_with_degree</span><span class="p">(</span><span class="n">top_genes_with_degrees</span><span class="p">,</span> <span class="n">gene_info</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>4 input query terms found dup hits:	[(&#39;IGHJ4&#39;, 2), (&#39;IGKV3-20&#39;, 2), (&#39;IGKV3-11&#39;, 2), (&#39;IGKV1-39&#39;, 2)]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene Symbol: IGKJ1
Degree: 118
Gene Name: immunoglobulin kappa joining 1
Entrez ID: 28950
Summary: Predicted to be involved in adaptive immune response. Predicted to be located in extracellular region and plasma membrane. Predicted to be part of immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: MZB1
Degree: 109
Gene Name: marginal zone B and B1 cell specific protein
Entrez ID: 51237
Summary: Involved in positive regulation of cell population proliferation. Located in cytoplasm and extracellular region. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: FCRL5
Degree: 108
Gene Name: Fc receptor like 5
Entrez ID: 83416
Summary: This gene encodes a member of the immunoglobulin receptor superfamily and the Fc-receptor like family. This gene and several other Fc receptor-like gene members are clustered on the long arm of chromosome 1. The encoded protein is a single-pass type I membrane protein and contains 8 immunoglobulin-like C2-type domains. This gene is implicated in B cell development and lymphomagenesis. Alternatively spliced transcript variants encoding different isoforms have been identified. [provided by RefSeq, Sep 2010].
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKC
Degree: 106
Gene Name: immunoglobulin kappa constant
Entrez ID: 3514
Summary: Predicted to enable antigen binding activity and immunoglobulin receptor binding activity. Involved in retina homeostasis. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGHJ4
Degree: 105
Gene Name: immunoglobulin heavy joining 4
Entrez ID: 28477
Summary: N/A
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGLL5
Degree: 99
Gene Name: immunoglobulin lambda like polypeptide 5
Entrez ID: 100423062
Summary: This gene encodes one of the immunoglobulin lambda-like polypeptides. It is located within the immunoglobulin lambda locus but it does not require somatic rearrangement for expression. The first exon of this gene is unrelated to immunoglobulin variable genes; the second and third exons are the immunoglobulin lambda joining 1 and the immunoglobulin lambda constant 1 gene segments. Alternative splicing results in multiple transcript variants. [provided by RefSeq, May 2010].
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV3-20
Degree: 97
Gene Name: immunoglobulin kappa variable 3-20
Entrez ID: 28912
Summary: Involved in antibacterial humoral response and glomerular filtration. Located in extracellular space. Part of monomeric IgA immunoglobulin complex; pentameric IgM immunoglobulin complex; and secretory IgA immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV3-11
Degree: 96
Gene Name: immunoglobulin kappa variable 3-11
Entrez ID: 28914
Summary: Predicted to be involved in immune response. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV1-39
Degree: 95
Gene Name: immunoglobulin kappa variable 1-39
Entrez ID: 28930
Summary: Predicted to be involved in immune response. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV3D-11
Degree: 95
Gene Name: immunoglobulin kappa variable 3D-11
Entrez ID: 28876
Summary: Predicted to be involved in immune response. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">KMeans</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># Assuming pearson_graph_pruned is your graph object</span>
<span class="n">adj_matrix</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">to_numpy_array</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>

<span class="c1"># Verify the shape of the adjacency matrix</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Adjacency matrix shape: </span><span class="si">{</span><span class="n">adj_matrix</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Number of clusters</span>
<span class="n">k</span> <span class="o">=</span> <span class="mi">6</span>

<span class="c1"># Apply K-means clustering</span>
<span class="n">kmeans</span> <span class="o">=</span> <span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">kmeans</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adj_matrix</span><span class="p">)</span>

<span class="c1"># Get cluster labels for each node</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">kmeans</span><span class="o">.</span><span class="n">labels_</span>

<span class="c1"># Map nodes to their corresponding cluster labels</span>
<span class="n">node_to_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="p">:</span> <span class="n">index</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="o">.</span><span class="n">nodes</span><span class="p">())}</span>
<span class="n">index_to_node</span> <span class="o">=</span> <span class="p">{</span><span class="n">index</span><span class="p">:</span> <span class="n">node</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">node_to_index</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

<span class="c1"># Print cluster labels for each node</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Cluster labels for each node:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">label</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">labels</span><span class="p">):</span>
    <span class="n">node</span> <span class="o">=</span> <span class="n">index_to_node</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Node </span><span class="si">{</span><span class="n">node</span><span class="si">}</span><span class="s2">: Cluster </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Draw the graph with nodes colored by their cluster label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>

<span class="c1"># Assign colors to nodes based on their cluster label</span>
<span class="n">node_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">labels</span><span class="p">[</span><span class="n">node_to_index</span><span class="p">[</span><span class="n">node</span><span class="p">]]</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">pearson_graph_pruned</span><span class="o">.</span><span class="n">nodes</span><span class="p">()]</span>

<span class="c1"># Draw the graph</span>
<span class="n">nx</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">with_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">node_color</span><span class="o">=</span><span class="n">node_colors</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">rainbow</span><span class="p">,</span> <span class="n">node_size</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">font_weight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">edge_color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Network Graph Visualization with K-means Clustering&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Adjacency matrix shape: (305, 305)
Cluster labels for each node:
Node SPN: Cluster 5
Node SCML4: Cluster 1
Node IL2RG: Cluster 1
Node CCL23: Cluster 1
Node TLR7: Cluster 5
Node FAM30A: Cluster 4
Node LILRB4: Cluster 1
Node IGKV4-1: Cluster 3
Node CXCR6: Cluster 1
Node FCGR1A: Cluster 1
Node FCGR3A: Cluster 5
Node GVINP1: Cluster 1
Node FASLG: Cluster 1
Node GBP5: Cluster 1
Node IGKV1ORY-1: Cluster 2
Node CLEC10A: Cluster 1
Node IGKV1-16: Cluster 0
Node GBP4: Cluster 1
Node IGHV4-55: Cluster 2
Node IGLV1-47: Cluster 0
Node IGHV1-67: Cluster 2
Node IGHV4-39: Cluster 1
Node CD22: Cluster 1
Node IL2RB: Cluster 1
Node CD300LF: Cluster 5
Node MSR1: Cluster 5
Node IGHV3-30: Cluster 0
Node IRF4: Cluster 1
Node IGLV2-8: Cluster 1
Node PDCD1LG2: Cluster 1
Node IGKV1D-42: Cluster 1
Node FCRL3: Cluster 1
Node IGHV4-59: Cluster 3
Node IGLV1-50: Cluster 1
Node CD74: Cluster 1
Node IGHV1OR15-9: Cluster 2
Node F13A1: Cluster 1
Node CD48: Cluster 1
Node P2RY13: Cluster 5
Node CRTAM: Cluster 1
Node CXCR3: Cluster 1
Node IGHJ3: Cluster 2
Node CD163: Cluster 1
Node IGHA1: Cluster 1
Node IGHV1OR15-2: Cluster 0
Node FCRL4: Cluster 1
Node THEMIS: Cluster 1
Node CXCL10: Cluster 1
Node ZNF683: Cluster 1
Node HLA-DQA1: Cluster 1
Node CD3D: Cluster 1
Node MS4A4A: Cluster 5
Node PTAFR: Cluster 1
Node PSMB9: Cluster 1
Node IGKV1D-8: Cluster 2
Node TBC1D27P: Cluster 1
Node HLA-DOA: Cluster 1
Node GZMB: Cluster 1
Node IGLV3-21: Cluster 0
Node MNDA: Cluster 5
Node FCRL1: Cluster 1
Node HLA-DQB1: Cluster 1
Node TIGIT: Cluster 1
Node IGHJ4: Cluster 3
Node IGHV3-21: Cluster 3
Node C1QA: Cluster 5
Node IGHV3OR15-7: Cluster 1
Node IGHV3OR16-8: Cluster 2
Node TRBC2: Cluster 1
Node PDCD1: Cluster 1
Node CXCL11: Cluster 1
Node ANKRD36BP2: Cluster 4
Node IGHV3-35: Cluster 1
Node IGKV1-27: Cluster 1
Node P2RX7: Cluster 1
Node TRBC1: Cluster 1
Node ITK: Cluster 1
Node SPIB: Cluster 1
Node IGLL5: Cluster 3
Node IGHV3-15: Cluster 0
Node TIFAB: Cluster 1
Node ALOX5: Cluster 1
Node IGKC: Cluster 3
Node IGHV4-34: Cluster 0
Node IGLV4-69: Cluster 1
Node LY86: Cluster 5
Node KLHL6: Cluster 1
Node SLC1A3: Cluster 1
Node OSCAR: Cluster 5
Node LINC00861: Cluster 1
Node PNOC: Cluster 4
Node GPR174: Cluster 1
Node IGSF6: Cluster 5
Node IGHJ5: Cluster 4
Node IGHGP: Cluster 1
Node CTLA4: Cluster 1
Node FOLR2: Cluster 1
Node IGHJ3P: Cluster 4
Node IGLV3-25: Cluster 0
Node IGLV1-44: Cluster 0
Node ALOX5AP: Cluster 1
Node FGL2: Cluster 1
Node TRAT1: Cluster 1
Node TREM2: Cluster 1
Node GBP1P1: Cluster 1
Node FCER2: Cluster 1
Node IGHV3-47: Cluster 1
Node NKG7: Cluster 1
Node IGHV2-26: Cluster 1
Node IGKV1OR2-3: Cluster 1
Node IGHG3: Cluster 0
Node IGHV3-62: Cluster 2
Node HLA-DRB1: Cluster 1
Node TLR8: Cluster 5
Node IGLV6-57: Cluster 0
Node C1QC: Cluster 5
Node IGHV3-7: Cluster 0
Node IGLV1-40: Cluster 0
Node IGHV4-61: Cluster 1
Node APOC1: Cluster 1
Node P2RY12: Cluster 1
Node LCK: Cluster 1
Node NIBAN3: Cluster 1
Node CD79B: Cluster 1
Node IGKV2OR22-4: Cluster 1
Node IGHV3-71: Cluster 2
Node IGHV5-51: Cluster 3
Node CD8A: Cluster 1
Node KLRB1: Cluster 1
Node IGKV3OR2-268: Cluster 2
Node JCHAIN: Cluster 1
Node IGHA2: Cluster 1
Node CCL19: Cluster 1
Node C1QB: Cluster 5
Node IGHV2-5: Cluster 0
Node IGKV1OR2-6: Cluster 2
Node CXCL9: Cluster 1
Node BCL2A1: Cluster 1
Node CPNE5: Cluster 4
Node IGKV1D-39: Cluster 2
Node IGLV7-43: Cluster 1
Node PELATON: Cluster 1
Node MARCO: Cluster 1
Node IGHV3-63: Cluster 1
Node IGHV3-6: Cluster 1
Node MZB1: Cluster 3
Node IGHV1-46: Cluster 0
Node SIRPG: Cluster 1
Node CALHM6: Cluster 1
Node IGHV3-33: Cluster 0
Node IGKV1OR22-5: Cluster 4
Node MRC1: Cluster 5
Node IGHV3-11: Cluster 0
Node CCL5: Cluster 1
Node CD8B: Cluster 1
Node C5AR1: Cluster 1
Node CD52: Cluster 1
Node CLEC17A: Cluster 1
Node HLA-DPA1: Cluster 1
Node IGKV1OR2-108: Cluster 2
Node IGHJ2: Cluster 2
Node IGKV1-9: Cluster 0
Node HLA-DPB1: Cluster 1
Node IGKV1-17: Cluster 1
Node IGHV3OR16-6: Cluster 1
Node IGKV1-8: Cluster 1
Node IGKV2-28: Cluster 1
Node IGKV3-20: Cluster 3
Node GIMAP5: Cluster 1
Node CD2: Cluster 1
Node LINC00926: Cluster 1
Node IGHV3OR16-9: Cluster 2
Node KLRD1: Cluster 1
Node IGKV3-7: Cluster 2
Node CD300C: Cluster 5
Node HLA-DRA: Cluster 1
Node IGKV1-33: Cluster 2
Node HLA-DRB5: Cluster 1
Node DPEP2: Cluster 1
Node BLK: Cluster 1
Node MEI1: Cluster 1
Node IGLV2-14: Cluster 0
Node IGHG1: Cluster 0
Node IGKV1OR2-11: Cluster 1
Node GPR34: Cluster 1
Node LINC02446: Cluster 1
Node IGHV3-23: Cluster 3
Node SLAMF6: Cluster 1
Node IGKV3-11: Cluster 3
Node CD3G: Cluster 1
Node IGHV6-1: Cluster 1
Node SLCO2B1: Cluster 5
Node IGKJ5: Cluster 4
Node IGHG2: Cluster 0
Node IGLV3-19: Cluster 0
Node FDCSP: Cluster 1
Node IGHV1-24: Cluster 0
Node SCIMP: Cluster 5
Node IGHV4-28: Cluster 3
Node CCL3: Cluster 1
Node GZMA: Cluster 1
Node TMEM150B: Cluster 1
Node LINC01781: Cluster 1
Node CCR2: Cluster 1
Node IGKJ4: Cluster 4
Node IGHV3-38: Cluster 1
Node IGKV3D-15: Cluster 2
Node PPP1R16B: Cluster 1
Node ICOS: Cluster 1
Node CD27: Cluster 4
Node DERL3: Cluster 2
Node IGKV1D-33: Cluster 2
Node IGHV1-12: Cluster 1
Node PTPRO: Cluster 5
Node IGHV3-52: Cluster 1
Node CCR7: Cluster 1
Node DOK2: Cluster 5
Node IGHV3-60: Cluster 1
Node IGHV3-74: Cluster 0
Node CNR2: Cluster 1
Node P2RY10: Cluster 1
Node IGKV1-39: Cluster 3
Node CD96: Cluster 1
Node IGKV3D-20: Cluster 3
Node IGHV3-53: Cluster 1
Node IGLV2-5: Cluster 0
Node IGLV3-1: Cluster 0
Node ITGB2: Cluster 1
Node ITGAM: Cluster 5
Node SH2D1A: Cluster 1
Node IGHV3-69-1: Cluster 2
Node CD19: Cluster 1
Node IGHV3-49: Cluster 0
Node LST1: Cluster 5
Node IGKJ1: Cluster 3
Node CD79A: Cluster 0
Node NUGGC: Cluster 4
Node GZMK: Cluster 1
Node FCRL5: Cluster 3
Node IFNG: Cluster 1
Node TMEM236: Cluster 1
Node PTPRC: Cluster 1
Node MS4A1: Cluster 1
Node IGHV3OR16-10: Cluster 1
Node IGHV3OR16-17: Cluster 2
Node PRAM1: Cluster 1
Node TFEC: Cluster 5
Node ZAP70: Cluster 1
Node IGHV1OR21-1: Cluster 1
Node IGKV1OR22-1: Cluster 1
Node IGHV4-30-2: Cluster 1
Node GZMH: Cluster 1
Node TRAC: Cluster 1
Node MAP3K5-AS2: Cluster 5
Node TYROBP: Cluster 5
Node IL7R: Cluster 1
Node ZNF831: Cluster 1
Node VPREB3: Cluster 1
Node IGKV2-30: Cluster 0
Node IGHV3-48: Cluster 0
Node IGKV1-6: Cluster 3
Node CCL4: Cluster 1
Node JAML: Cluster 5
Node LILRA6: Cluster 1
Node IGKV3-15: Cluster 3
Node IGKV3D-11: Cluster 3
Node SLAMF8: Cluster 1
Node IGKV1-5: Cluster 3
Node IGHJ1: Cluster 1
Node IGLV2-11: Cluster 0
Node TNFRSF13C: Cluster 1
Node PLEK: Cluster 5
Node IGLV2-23: Cluster 0
Node POU2AF1: Cluster 4
Node IGLV1-51: Cluster 1
Node FGR: Cluster 5
Node HCK: Cluster 5
Node RTN1: Cluster 1
Node IGHV1-18: Cluster 3
Node IGKV2D-28: Cluster 1
Node IGKV1-12: Cluster 0
Node IGKV2D-30: Cluster 1
Node TNFRSF13B: Cluster 1
Node STX11: Cluster 1
Node LINC02362: Cluster 4
Node IGHV3-41: Cluster 1
Node IGHV4-80: Cluster 1
Node FCRLA: Cluster 1
Node HK3: Cluster 5
Node VSIG4: Cluster 5
Node EOMES: Cluster 1
Node TLR10: Cluster 1
Node TNFRSF17: Cluster 3
Node IGLC3: Cluster 0
Node IGHV3-19: Cluster 1
Node GBP1: Cluster 1
Node IGHV1-69D: Cluster 0
Node LAX1: Cluster 4
Node IGKV1D-12: Cluster 1
Node IGHV3OR16-13: Cluster 2
Node CD3E: Cluster 1
Node IGLC2: Cluster 3
Node SIT1: Cluster 1
Node MS4A7: Cluster 5
Node CYBB: Cluster 5
</pre></div>
</div>
<img alt="../_images/32c1fa6381d5968c2e0fd042088ff7843168c03f58b90fde064759267c3ec094.png" src="../_images/32c1fa6381d5968c2e0fd042088ff7843168c03f58b90fde064759267c3ec094.png" />
</div>
</div>
</section>
<section id="interactive-visualization">
<h1>Interactive Visualization<a class="headerlink" href="#interactive-visualization" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">visualize_interactive_network</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;network.html&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes a NetworkX graph interactively using PyVis.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The NetworkX graph to visualize.</span>
<span class="sd">    output_file (str): The filename for the output HTML file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">net</span> <span class="o">=</span> <span class="n">Network</span><span class="p">(</span><span class="n">notebook</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="s2">&quot;100%&quot;</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="s2">&quot;1000px&quot;</span><span class="p">)</span>
    <span class="n">net</span><span class="o">.</span><span class="n">from_nx</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
    <span class="n">net</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the graph interactively</span>
<span class="n">visualize_interactive_network</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">intermediate_data_dir</span><span class="p">,</span> <span class="s1">&#39;pearson_network.html&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">save_interactive_network_plotly</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;network.html&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Interactive Network&quot;</span><span class="p">):</span>
    <span class="c1"># Extract node positions</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>

    <span class="c1"># Extract edges</span>
    <span class="n">edge_x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">edge_y</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
        <span class="n">x0</span><span class="p">,</span> <span class="n">y0</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">edge</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
        <span class="n">x1</span><span class="p">,</span> <span class="n">y1</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">edge</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">edge_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x0</span><span class="p">)</span>
        <span class="n">edge_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x1</span><span class="p">)</span>
        <span class="n">edge_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">edge_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y0</span><span class="p">)</span>
        <span class="n">edge_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y1</span><span class="p">)</span>
        <span class="n">edge_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="n">edge_trace</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">edge_x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">edge_y</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#888&#39;</span><span class="p">),</span>
        <span class="n">hoverinfo</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines&#39;</span><span class="p">)</span>

    <span class="c1"># Extract nodes</span>
    <span class="n">node_x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">node_y</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">node_text</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">node</span><span class="p">]</span>
        <span class="n">node_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">node_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
        <span class="n">node_text</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="p">))</span>

    <span class="n">node_trace</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">node_x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">node_y</span><span class="p">,</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers+text&#39;</span><span class="p">,</span>
        <span class="n">text</span><span class="o">=</span><span class="n">node_text</span><span class="p">,</span>
        <span class="n">hoverinfo</span><span class="o">=</span><span class="s1">&#39;text&#39;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
            <span class="n">showscale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">colorscale</span><span class="o">=</span><span class="s1">&#39;YlGnBu&#39;</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
            <span class="n">colorbar</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">thickness</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Node Connections&#39;</span><span class="p">,</span>
                <span class="n">xanchor</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span>
                <span class="n">titleside</span><span class="o">=</span><span class="s1">&#39;right&#39;</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[</span><span class="n">edge_trace</span><span class="p">,</span> <span class="n">node_trace</span><span class="p">],</span>
                    <span class="n">layout</span><span class="o">=</span><span class="n">go</span><span class="o">.</span><span class="n">Layout</span><span class="p">(</span>
                        <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
                        <span class="n">titlefont_size</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                        <span class="n">showlegend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="n">hovermode</span><span class="o">=</span><span class="s1">&#39;closest&#39;</span><span class="p">,</span>
                        <span class="n">margin</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">b</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">l</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">r</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">t</span><span class="o">=</span><span class="mi">40</span><span class="p">),</span>
                        <span class="n">annotations</span><span class="o">=</span><span class="p">[</span> <span class="nb">dict</span><span class="p">(</span>
                            <span class="n">text</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                            <span class="n">showarrow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                            <span class="n">xref</span><span class="o">=</span><span class="s2">&quot;paper&quot;</span><span class="p">,</span> <span class="n">yref</span><span class="o">=</span><span class="s2">&quot;paper&quot;</span>
                        <span class="p">)],</span>
                        <span class="n">xaxis</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">zeroline</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
                        <span class="n">yaxis</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">zeroline</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
                    <span class="p">)</span>

    <span class="c1"># Save the figure to an HTML file</span>
    <span class="n">pio</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">output_file</span><span class="p">,</span> <span class="n">auto_open</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="todo-add-hover-info-to-plotly-plot-gene-info">
<h1>TODO: add hover info to plotly plot (gene info)<a class="headerlink" href="#todo-add-hover-info-to-plotly-plot-gene-info" title="Permalink to this heading">#</a></h1>
</section>
<section id="todo-introduce-clustering-kmeans-etc">
<h1>TODO: introduce clustering kmeans. etc<a class="headerlink" href="#todo-introduce-clustering-kmeans-etc" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">save_interactive_network_plotly</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;genecoexp_plotly_network.html&#39;</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Gene Coexpression Network&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="storing-generated-networks">
<h1>Storing generated networks<a class="headerlink" href="#storing-generated-networks" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#example</span>
<span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;gene_coexpression_network.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="patient-similarity-network">
<h1>Patient similarity network<a class="headerlink" href="#patient-similarity-network" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Based on the same expression matrix we can create a patient similarity network.</span>
<span class="c1"># By transposing the matrix will switch the rows and columns,</span>
<span class="c1"># meaning that patients will become the rows and genes will become the columns.</span>
<span class="c1"># By doing this, you can compute the correlation (or similarity) between patients based on their gene expression profiles,</span>
<span class="c1"># and then create a network where nodes represent patients and edges represent similarities.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_renamed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;expression_data_filtered.csv&quot;</span><span class="p">),</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Transpose the matrix</span>
<span class="n">patient_gene_matrix</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_gene_matrix</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TCGA-38-7271</th>
      <th>TCGA-55-7914</th>
      <th>TCGA-95-7043</th>
      <th>TCGA-73-4658</th>
      <th>TCGA-86-8076</th>
      <th>TCGA-55-7726</th>
      <th>TCGA-44-6147</th>
      <th>TCGA-50-5932</th>
      <th>TCGA-44-2661</th>
      <th>TCGA-86-7954</th>
      <th>...</th>
      <th>TCGA-97-A4M7</th>
      <th>TCGA-62-A46R</th>
      <th>TCGA-50-5055</th>
      <th>TCGA-38-4628</th>
      <th>TCGA-86-7713</th>
      <th>TCGA-86-8073</th>
      <th>TCGA-MN-A4N4</th>
      <th>TCGA-53-7626</th>
      <th>TCGA-44-A47G</th>
      <th>TCGA-55-6969</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A2M</th>
      <td>17.7492</td>
      <td>14.8513</td>
      <td>14.1691</td>
      <td>16.7238</td>
      <td>15.6783</td>
      <td>14.7566</td>
      <td>16.4368</td>
      <td>15.5476</td>
      <td>15.5478</td>
      <td>15.1337</td>
      <td>...</td>
      <td>15.9553</td>
      <td>13.9511</td>
      <td>16.3097</td>
      <td>14.3934</td>
      <td>15.8254</td>
      <td>16.3773</td>
      <td>14.9411</td>
      <td>16.7343</td>
      <td>15.6622</td>
      <td>14.8136</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>4.4411</td>
      <td>4.4530</td>
      <td>4.5026</td>
      <td>3.1704</td>
      <td>4.7422</td>
      <td>6.0918</td>
      <td>4.0602</td>
      <td>3.0827</td>
      <td>3.4608</td>
      <td>3.6367</td>
      <td>...</td>
      <td>4.7076</td>
      <td>2.4821</td>
      <td>8.0810</td>
      <td>3.3415</td>
      <td>5.3379</td>
      <td>4.0817</td>
      <td>6.7210</td>
      <td>5.2575</td>
      <td>3.7345</td>
      <td>4.3058</td>
    </tr>
    <tr>
      <th>A4GALT</th>
      <td>10.1862</td>
      <td>8.9312</td>
      <td>9.0834</td>
      <td>9.1443</td>
      <td>9.5150</td>
      <td>11.4452</td>
      <td>9.1462</td>
      <td>7.3597</td>
      <td>9.1602</td>
      <td>8.9344</td>
      <td>...</td>
      <td>9.1019</td>
      <td>8.6653</td>
      <td>10.8837</td>
      <td>8.6304</td>
      <td>9.3827</td>
      <td>9.4234</td>
      <td>11.9107</td>
      <td>9.5379</td>
      <td>9.8660</td>
      <td>9.7702</td>
    </tr>
    <tr>
      <th>AACSP1</th>
      <td>3.5845</td>
      <td>3.6762</td>
      <td>3.8623</td>
      <td>2.4821</td>
      <td>3.2271</td>
      <td>2.4821</td>
      <td>5.6614</td>
      <td>3.0827</td>
      <td>3.9428</td>
      <td>4.7089</td>
      <td>...</td>
      <td>5.4947</td>
      <td>3.0956</td>
      <td>2.4821</td>
      <td>3.7710</td>
      <td>2.9397</td>
      <td>6.6519</td>
      <td>8.8283</td>
      <td>2.4821</td>
      <td>4.1045</td>
      <td>6.8425</td>
    </tr>
    <tr>
      <th>AADAC</th>
      <td>9.6415</td>
      <td>6.5685</td>
      <td>5.5634</td>
      <td>4.3206</td>
      <td>7.2321</td>
      <td>6.8821</td>
      <td>8.1176</td>
      <td>4.7935</td>
      <td>11.7750</td>
      <td>7.0644</td>
      <td>...</td>
      <td>9.0856</td>
      <td>3.8162</td>
      <td>5.6471</td>
      <td>4.6785</td>
      <td>4.6578</td>
      <td>6.7641</td>
      <td>3.4947</td>
      <td>5.7120</td>
      <td>6.2214</td>
      <td>5.2016</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>ZSCAN23</th>
      <td>5.7385</td>
      <td>4.3796</td>
      <td>5.6728</td>
      <td>3.6533</td>
      <td>5.0478</td>
      <td>6.0489</td>
      <td>5.8058</td>
      <td>3.7897</td>
      <td>4.3451</td>
      <td>5.5259</td>
      <td>...</td>
      <td>4.7076</td>
      <td>2.4821</td>
      <td>4.6683</td>
      <td>3.4697</td>
      <td>4.3516</td>
      <td>7.2074</td>
      <td>6.9167</td>
      <td>6.0927</td>
      <td>5.3854</td>
      <td>3.6792</td>
    </tr>
    <tr>
      <th>ZSCAN31</th>
      <td>9.0817</td>
      <td>10.9565</td>
      <td>10.8095</td>
      <td>8.9874</td>
      <td>9.5810</td>
      <td>9.0185</td>
      <td>10.2085</td>
      <td>9.8803</td>
      <td>11.0327</td>
      <td>10.2707</td>
      <td>...</td>
      <td>9.2033</td>
      <td>10.5020</td>
      <td>8.8073</td>
      <td>9.8672</td>
      <td>10.6419</td>
      <td>10.2410</td>
      <td>9.4672</td>
      <td>10.0507</td>
      <td>7.4843</td>
      <td>8.5268</td>
    </tr>
    <tr>
      <th>ZSCAN4</th>
      <td>3.2709</td>
      <td>3.0918</td>
      <td>2.4821</td>
      <td>3.4468</td>
      <td>3.0118</td>
      <td>5.0894</td>
      <td>5.5341</td>
      <td>3.5082</td>
      <td>5.4987</td>
      <td>3.3645</td>
      <td>...</td>
      <td>4.4088</td>
      <td>7.4937</td>
      <td>4.3167</td>
      <td>5.0907</td>
      <td>6.7676</td>
      <td>5.1126</td>
      <td>3.3143</td>
      <td>4.2816</td>
      <td>4.4650</td>
      <td>2.4821</td>
    </tr>
    <tr>
      <th>ZWINT</th>
      <td>9.2615</td>
      <td>10.1334</td>
      <td>10.9149</td>
      <td>9.7828</td>
      <td>8.6640</td>
      <td>11.1212</td>
      <td>9.0718</td>
      <td>10.7305</td>
      <td>9.0013</td>
      <td>10.4268</td>
      <td>...</td>
      <td>9.7614</td>
      <td>10.7593</td>
      <td>9.7087</td>
      <td>10.4729</td>
      <td>10.9794</td>
      <td>10.1390</td>
      <td>10.0583</td>
      <td>9.4627</td>
      <td>9.5817</td>
      <td>10.8228</td>
    </tr>
    <tr>
      <th>ZYG11A</th>
      <td>4.0073</td>
      <td>8.3219</td>
      <td>6.7593</td>
      <td>3.4468</td>
      <td>4.7844</td>
      <td>5.0894</td>
      <td>9.4014</td>
      <td>4.7935</td>
      <td>4.2748</td>
      <td>7.1709</td>
      <td>...</td>
      <td>6.5412</td>
      <td>7.1933</td>
      <td>5.8778</td>
      <td>7.3822</td>
      <td>8.7596</td>
      <td>8.2478</td>
      <td>8.4566</td>
      <td>7.4014</td>
      <td>4.6804</td>
      <td>8.6068</td>
    </tr>
  </tbody>
</table>
<p>5338 rows Ã— 498 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dictionary to store different correlation matrices</span>
<span class="n">patient_correlation_matrices</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Pearson correlation</span>
<span class="n">patient_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">patient_gene_matrix</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>

<span class="c1"># patient_correlation_matrices[&#39;biweight_midcorrelation&#39;] = abs_bicorr(patient_gene_matrix)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_pearson_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">patient_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">verify_edge_weights</span><span class="p">(</span><span class="n">patient_pearson_graph</span><span class="p">,</span> <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample of 10 edges with weights:
Edge (TCGA-38-7271, TCGA-73-4658) - weight: 0.8696511707884843
Edge (TCGA-38-7271, TCGA-86-8076) - weight: 0.8173779911639248
Edge (TCGA-38-7271, TCGA-44-2661) - weight: 0.8652686243802332
Edge (TCGA-38-7271, TCGA-73-4662) - weight: 0.8446135021161085
Edge (TCGA-38-7271, TCGA-55-6986) - weight: 0.833876175940164
Edge (TCGA-38-7271, TCGA-49-6744) - weight: 0.8919809060862212
Edge (TCGA-38-7271, TCGA-69-7763) - weight: 0.8374050473870502
Edge (TCGA-38-7271, TCGA-44-6774) - weight: 0.802120255633156
Edge (TCGA-38-7271, TCGA-67-3774) - weight: 0.8179727328277505
Edge (TCGA-38-7271, TCGA-97-A4M2) - weight: 0.8045412817214057
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">patient_pearson_graph</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/0da47457af7538f888c711b78dc4c47e6e034413d8570f3f548aa02a86ec6be9.png" src="../_images/0da47457af7538f888c711b78dc4c47e6e034413d8570f3f548aa02a86ec6be9.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">patient_pearson_graph_pruned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">patient_pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Patient Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1a56d12e97838d957acd64e46d274910d979d49f815c6afa677e51e08637e578.png" src="../_images/1a56d12e97838d957acd64e46d274910d979d49f815c6afa677e51e08637e578.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_pearson_graph_pruned_knn</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 434
Number of edges: 10373
Sample nodes: [&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7914&#39;, &#39;TCGA-73-4658&#39;, &#39;TCGA-86-8076&#39;, &#39;TCGA-55-7726&#39;, &#39;TCGA-44-6147&#39;, &#39;TCGA-50-5932&#39;, &#39;TCGA-44-2661&#39;, &#39;TCGA-86-7954&#39;, &#39;TCGA-73-4662&#39;]
Sample edges: [(&#39;TCGA-38-7271&#39;, &#39;TCGA-73-4658&#39;, {&#39;weight&#39;: 0.8696511707884843}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-86-8076&#39;, {&#39;weight&#39;: 0.8173779911639248}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-44-2661&#39;, {&#39;weight&#39;: 0.8652686243802332}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-73-4662&#39;, {&#39;weight&#39;: 0.8446135021161085}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-55-6986&#39;, {&#39;weight&#39;: 0.833876175940164}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-49-6744&#39;, {&#39;weight&#39;: 0.8919809060862212}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-69-7763&#39;, {&#39;weight&#39;: 0.8374050473870502}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-44-6774&#39;, {&#39;weight&#39;: 0.802120255633156}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-67-3774&#39;, {&#39;weight&#39;: 0.8179727328277505}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-97-A4M2&#39;, {&#39;weight&#39;: 0.8045412817214057})]
No self-loops in the graph.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 434
Number of edges: 2956
Sample nodes: [&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7914&#39;, &#39;TCGA-73-4658&#39;, &#39;TCGA-86-8076&#39;, &#39;TCGA-55-7726&#39;, &#39;TCGA-44-6147&#39;, &#39;TCGA-50-5932&#39;, &#39;TCGA-44-2661&#39;, &#39;TCGA-86-7954&#39;, &#39;TCGA-73-4662&#39;]
Sample edges: [(&#39;TCGA-38-7271&#39;, &#39;TCGA-53-7626&#39;, {&#39;weight&#39;: 0.8932510135537189}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-49-6744&#39;, {&#39;weight&#39;: 0.8919809060862212}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-78-8648&#39;, {&#39;weight&#39;: 0.8857601471754217}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-86-8671&#39;, {&#39;weight&#39;: 0.8825367028721114}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-50-5941&#39;, {&#39;weight&#39;: 0.8814255405956031}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7574&#39;, {&#39;weight&#39;: 0.8799541588637929}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-97-7553&#39;, {&#39;weight&#39;: 0.8775024097687647}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-50-5055&#39;, {&#39;weight&#39;: 0.8726819735362153}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-50-5045&#39;, {&#39;weight&#39;: 0.8723464616125998}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-73-4658&#39;, {&#39;weight&#39;: 0.8696511707884843})]
No self-loops in the graph.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;K-Nearest Neighbors (k=10) Patient Correlation Network&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a31c90769b5f745cd8f3a2725aa466383582baccccc34dab762d10b98c50a01c.png" src="../_images/a31c90769b5f745cd8f3a2725aa466383582baccccc34dab762d10b98c50a01c.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;patient_coexpression_network_knn.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="dnameth-pn">
<h1>DNAmeth PN<a class="headerlink" href="#dnameth-pn" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the data using pickle</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;ISMB_TCGA_DNAm.pkl&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span> <span class="p">:</span> 
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

<span class="n">meth_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;datExpr&quot;</span><span class="p">]</span>
<span class="n">meth_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cg23057992</th>
      <th>cg16602369</th>
      <th>cg20545544</th>
      <th>cg04571941</th>
      <th>cg23091104</th>
      <th>cg11738485</th>
      <th>cg17735539</th>
      <th>cg12662576</th>
      <th>cg10554839</th>
      <th>cg08866780</th>
      <th>...</th>
      <th>cg04339424</th>
      <th>cg01318691</th>
      <th>cg27068965</th>
      <th>cg23028848</th>
      <th>cg21122529</th>
      <th>cg22254072</th>
      <th>cg10237911</th>
      <th>cg09874052</th>
      <th>cg27624178</th>
      <th>cg02857943</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-55-7914</th>
      <td>0.039144</td>
      <td>0.078881</td>
      <td>0.032201</td>
      <td>0.098520</td>
      <td>0.971626</td>
      <td>0.020020</td>
      <td>0.929670</td>
      <td>0.033325</td>
      <td>0.043767</td>
      <td>0.028051</td>
      <td>...</td>
      <td>0.053027</td>
      <td>0.046047</td>
      <td>0.951328</td>
      <td>0.045767</td>
      <td>0.048185</td>
      <td>0.014262</td>
      <td>0.034541</td>
      <td>0.034967</td>
      <td>0.041538</td>
      <td>0.053116</td>
    </tr>
    <tr>
      <th>TCGA-38-4631</th>
      <td>0.954805</td>
      <td>0.047287</td>
      <td>0.027976</td>
      <td>0.925794</td>
      <td>0.144636</td>
      <td>0.365842</td>
      <td>0.156414</td>
      <td>0.027917</td>
      <td>0.047609</td>
      <td>0.049901</td>
      <td>...</td>
      <td>0.048398</td>
      <td>0.047221</td>
      <td>0.939891</td>
      <td>0.056141</td>
      <td>0.047459</td>
      <td>0.014548</td>
      <td>0.056591</td>
      <td>0.039086</td>
      <td>0.044946</td>
      <td>0.055473</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>0.966071</td>
      <td>0.126549</td>
      <td>0.030316</td>
      <td>0.945166</td>
      <td>0.976080</td>
      <td>0.975638</td>
      <td>0.942018</td>
      <td>0.066659</td>
      <td>0.052100</td>
      <td>0.047054</td>
      <td>...</td>
      <td>0.040815</td>
      <td>0.042040</td>
      <td>0.951235</td>
      <td>0.045477</td>
      <td>0.074037</td>
      <td>0.019053</td>
      <td>0.029978</td>
      <td>0.039345</td>
      <td>0.060650</td>
      <td>0.041385</td>
    </tr>
    <tr>
      <th>TCGA-50-5932</th>
      <td>0.965140</td>
      <td>0.061194</td>
      <td>0.935457</td>
      <td>0.923421</td>
      <td>0.965443</td>
      <td>0.027598</td>
      <td>0.040885</td>
      <td>0.036178</td>
      <td>0.045745</td>
      <td>0.046439</td>
      <td>...</td>
      <td>0.057840</td>
      <td>0.045106</td>
      <td>0.948255</td>
      <td>0.044558</td>
      <td>0.039484</td>
      <td>0.013837</td>
      <td>0.037048</td>
      <td>0.032503</td>
      <td>0.058576</td>
      <td>0.037666</td>
    </tr>
    <tr>
      <th>TCGA-55-7576</th>
      <td>0.958612</td>
      <td>0.071607</td>
      <td>0.027077</td>
      <td>0.940574</td>
      <td>0.976870</td>
      <td>0.785073</td>
      <td>0.953674</td>
      <td>0.945107</td>
      <td>0.056088</td>
      <td>0.955472</td>
      <td>...</td>
      <td>0.046921</td>
      <td>0.051442</td>
      <td>0.954000</td>
      <td>0.034352</td>
      <td>0.031497</td>
      <td>0.014924</td>
      <td>0.035520</td>
      <td>0.043547</td>
      <td>0.056388</td>
      <td>0.052208</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-50-8457</th>
      <td>0.960316</td>
      <td>0.940219</td>
      <td>0.046927</td>
      <td>0.942980</td>
      <td>0.977751</td>
      <td>0.633731</td>
      <td>0.940723</td>
      <td>0.053398</td>
      <td>0.059292</td>
      <td>0.968875</td>
      <td>...</td>
      <td>0.047640</td>
      <td>0.062680</td>
      <td>0.954824</td>
      <td>0.043137</td>
      <td>0.048599</td>
      <td>0.014066</td>
      <td>0.031109</td>
      <td>0.039247</td>
      <td>0.035110</td>
      <td>0.054273</td>
    </tr>
    <tr>
      <th>TCGA-91-6840</th>
      <td>0.964250</td>
      <td>0.921250</td>
      <td>0.957943</td>
      <td>0.958416</td>
      <td>0.981439</td>
      <td>0.278432</td>
      <td>0.968525</td>
      <td>0.032143</td>
      <td>0.083996</td>
      <td>0.037912</td>
      <td>...</td>
      <td>0.045197</td>
      <td>0.066271</td>
      <td>0.963420</td>
      <td>0.047621</td>
      <td>0.060854</td>
      <td>0.015823</td>
      <td>0.045674</td>
      <td>0.040006</td>
      <td>0.052008</td>
      <td>0.040826</td>
    </tr>
    <tr>
      <th>TCGA-86-8280</th>
      <td>0.948855</td>
      <td>0.942925</td>
      <td>0.958506</td>
      <td>0.945667</td>
      <td>0.970511</td>
      <td>0.544447</td>
      <td>0.921924</td>
      <td>0.939427</td>
      <td>0.082121</td>
      <td>0.045036</td>
      <td>...</td>
      <td>0.062367</td>
      <td>0.058898</td>
      <td>0.954355</td>
      <td>0.045702</td>
      <td>0.048930</td>
      <td>0.019321</td>
      <td>0.039341</td>
      <td>0.036627</td>
      <td>0.054269</td>
      <td>0.056795</td>
    </tr>
    <tr>
      <th>TCGA-95-A4VK</th>
      <td>0.037255</td>
      <td>0.050940</td>
      <td>0.036565</td>
      <td>0.098237</td>
      <td>0.976780</td>
      <td>0.683394</td>
      <td>0.960073</td>
      <td>0.025232</td>
      <td>0.062487</td>
      <td>0.026908</td>
      <td>...</td>
      <td>0.056248</td>
      <td>0.053591</td>
      <td>0.947020</td>
      <td>0.035344</td>
      <td>0.057783</td>
      <td>0.010943</td>
      <td>0.051868</td>
      <td>0.035566</td>
      <td>0.044910</td>
      <td>0.042269</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>0.945954</td>
      <td>0.068404</td>
      <td>0.923325</td>
      <td>0.930684</td>
      <td>0.962295</td>
      <td>0.037952</td>
      <td>0.153943</td>
      <td>0.033980</td>
      <td>0.056811</td>
      <td>0.050211</td>
      <td>...</td>
      <td>0.048765</td>
      <td>0.061228</td>
      <td>0.930596</td>
      <td>0.056050</td>
      <td>0.061802</td>
      <td>0.017861</td>
      <td>0.044229</td>
      <td>0.054682</td>
      <td>0.045090</td>
      <td>0.064186</td>
    </tr>
  </tbody>
</table>
<p>459 rows Ã— 300000 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">smoking_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;smoking.tsv&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="c1">#Tab seperated document from EWAS</span>
<span class="n">common_annotated_cpgs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">smoking_df</span><span class="p">[</span><span class="s1">&#39;cpg&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()[</span><span class="n">smoking_df</span><span class="p">[</span><span class="s1">&#39;cpg&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> 
<span class="n">cpgs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">common_annotated_cpgs</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">meth_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>


<span class="n">common_patients</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">meth_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>


<span class="n">meth_data_filt</span> <span class="o">=</span> <span class="n">meth_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span> <span class="nb">list</span><span class="p">(</span><span class="n">common_patients</span><span class="p">)</span> <span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">cpgs</span><span class="p">)]</span>

<span class="n">patient_meth_matrix</span> <span class="o">=</span> <span class="n">meth_data_filt</span><span class="o">.</span><span class="n">T</span>


<span class="c1"># Keep only patients who are prevalent in the patient gene network</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">meth_data_filt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cg14712058</th>
      <th>cg07123182</th>
      <th>cg21913886</th>
      <th>cg08149865</th>
      <th>cg19717773</th>
      <th>cg01731783</th>
      <th>cg26718213</th>
      <th>cg17287155</th>
      <th>cg11660018</th>
      <th>cg00501876</th>
      <th>...</th>
      <th>cg23079012</th>
      <th>cg12876356</th>
      <th>cg01899089</th>
      <th>cg19089201</th>
      <th>cg02583484</th>
      <th>cg12806681</th>
      <th>cg17372101</th>
      <th>cg07251887</th>
      <th>cg06235438</th>
      <th>cg11902777</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-53-7813</th>
      <td>0.674583</td>
      <td>0.698905</td>
      <td>0.576494</td>
      <td>0.795106</td>
      <td>0.812521</td>
      <td>0.445949</td>
      <td>0.307335</td>
      <td>0.934581</td>
      <td>0.439352</td>
      <td>0.230338</td>
      <td>...</td>
      <td>0.967584</td>
      <td>0.402212</td>
      <td>0.378213</td>
      <td>0.825657</td>
      <td>0.218691</td>
      <td>0.895352</td>
      <td>0.327018</td>
      <td>0.411920</td>
      <td>0.771466</td>
      <td>0.147499</td>
    </tr>
    <tr>
      <th>TCGA-44-A479</th>
      <td>0.733725</td>
      <td>0.737029</td>
      <td>0.645343</td>
      <td>0.768585</td>
      <td>0.879891</td>
      <td>0.504076</td>
      <td>0.330575</td>
      <td>0.765683</td>
      <td>0.497904</td>
      <td>0.356694</td>
      <td>...</td>
      <td>0.966014</td>
      <td>0.542518</td>
      <td>0.482956</td>
      <td>0.863318</td>
      <td>0.277797</td>
      <td>0.856748</td>
      <td>0.341018</td>
      <td>0.612926</td>
      <td>0.603552</td>
      <td>0.251977</td>
    </tr>
    <tr>
      <th>TCGA-55-7911</th>
      <td>0.785928</td>
      <td>0.915015</td>
      <td>0.525344</td>
      <td>0.804578</td>
      <td>0.756804</td>
      <td>0.406378</td>
      <td>0.527881</td>
      <td>0.934787</td>
      <td>0.729670</td>
      <td>0.329801</td>
      <td>...</td>
      <td>0.972162</td>
      <td>0.640486</td>
      <td>0.406611</td>
      <td>0.814884</td>
      <td>0.253345</td>
      <td>0.910143</td>
      <td>0.286922</td>
      <td>0.392996</td>
      <td>0.626979</td>
      <td>0.149220</td>
    </tr>
    <tr>
      <th>TCGA-49-6761</th>
      <td>0.724054</td>
      <td>0.786670</td>
      <td>0.499649</td>
      <td>0.796624</td>
      <td>0.854948</td>
      <td>0.475507</td>
      <td>0.630280</td>
      <td>0.962896</td>
      <td>0.561377</td>
      <td>0.306555</td>
      <td>...</td>
      <td>0.959675</td>
      <td>0.390715</td>
      <td>0.578740</td>
      <td>0.862246</td>
      <td>0.294194</td>
      <td>0.915717</td>
      <td>0.358318</td>
      <td>0.407083</td>
      <td>0.779001</td>
      <td>0.452867</td>
    </tr>
    <tr>
      <th>TCGA-97-7547</th>
      <td>0.862960</td>
      <td>0.680225</td>
      <td>0.491567</td>
      <td>0.897165</td>
      <td>0.891053</td>
      <td>0.396081</td>
      <td>0.302026</td>
      <td>0.967135</td>
      <td>0.539988</td>
      <td>0.236335</td>
      <td>...</td>
      <td>0.971201</td>
      <td>0.307213</td>
      <td>0.457228</td>
      <td>0.823568</td>
      <td>0.293453</td>
      <td>0.927632</td>
      <td>0.729051</td>
      <td>0.405593</td>
      <td>0.835712</td>
      <td>0.652962</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-64-5779</th>
      <td>0.774646</td>
      <td>0.621952</td>
      <td>0.647204</td>
      <td>0.810046</td>
      <td>0.798557</td>
      <td>0.518443</td>
      <td>0.302897</td>
      <td>0.942683</td>
      <td>0.360692</td>
      <td>0.273212</td>
      <td>...</td>
      <td>0.981816</td>
      <td>0.416046</td>
      <td>0.537424</td>
      <td>0.850061</td>
      <td>0.574614</td>
      <td>0.946129</td>
      <td>0.251461</td>
      <td>0.403382</td>
      <td>0.826132</td>
      <td>0.304593</td>
    </tr>
    <tr>
      <th>TCGA-97-8552</th>
      <td>0.813098</td>
      <td>0.847171</td>
      <td>0.430333</td>
      <td>0.897129</td>
      <td>0.832132</td>
      <td>0.436413</td>
      <td>0.333738</td>
      <td>0.965335</td>
      <td>0.605481</td>
      <td>0.284973</td>
      <td>...</td>
      <td>0.968935</td>
      <td>0.366322</td>
      <td>0.490677</td>
      <td>0.714217</td>
      <td>0.546280</td>
      <td>0.933957</td>
      <td>0.757759</td>
      <td>0.426996</td>
      <td>0.799862</td>
      <td>0.411487</td>
    </tr>
    <tr>
      <th>TCGA-71-6725</th>
      <td>0.716992</td>
      <td>0.718337</td>
      <td>0.271827</td>
      <td>0.821352</td>
      <td>0.918169</td>
      <td>0.417347</td>
      <td>0.278869</td>
      <td>0.665987</td>
      <td>0.288103</td>
      <td>0.298375</td>
      <td>...</td>
      <td>0.967062</td>
      <td>0.271342</td>
      <td>0.266997</td>
      <td>0.370942</td>
      <td>0.316028</td>
      <td>0.882963</td>
      <td>0.435445</td>
      <td>0.272366</td>
      <td>0.693021</td>
      <td>0.105007</td>
    </tr>
    <tr>
      <th>TCGA-91-A4BC</th>
      <td>0.735027</td>
      <td>0.725628</td>
      <td>0.538211</td>
      <td>0.793583</td>
      <td>0.518587</td>
      <td>0.526517</td>
      <td>0.422005</td>
      <td>0.920201</td>
      <td>0.493812</td>
      <td>0.377786</td>
      <td>...</td>
      <td>0.961240</td>
      <td>0.533733</td>
      <td>0.495279</td>
      <td>0.809056</td>
      <td>0.263272</td>
      <td>0.933790</td>
      <td>0.296476</td>
      <td>0.553004</td>
      <td>0.654253</td>
      <td>0.306767</td>
    </tr>
    <tr>
      <th>TCGA-55-8615</th>
      <td>0.895974</td>
      <td>0.589359</td>
      <td>0.352932</td>
      <td>0.878058</td>
      <td>0.753100</td>
      <td>0.357509</td>
      <td>0.244239</td>
      <td>0.928896</td>
      <td>0.411669</td>
      <td>0.242435</td>
      <td>...</td>
      <td>0.973788</td>
      <td>0.302342</td>
      <td>0.298462</td>
      <td>0.879213</td>
      <td>0.177547</td>
      <td>0.910429</td>
      <td>0.185487</td>
      <td>0.256841</td>
      <td>0.845977</td>
      <td>0.141203</td>
    </tr>
  </tbody>
</table>
<p>385 rows Ã— 112 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_meth_matrix</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dictionary to store different correlation matrices</span>
<span class="n">p_meth_correlation_matrices</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Pearson correlation</span>
<span class="n">p_meth_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">patient_meth_matrix</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>
<span class="n">p_meth_pearson_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">p_meth_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
<span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">p_meth_pearson_graph_pruned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">p_meth_pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">visualize_graph</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1f3dc4d62a43490cbaac093a8fca1e18f8bc37f69e72f8396ac1f770f7aec8b9.png" src="../_images/1f3dc4d62a43490cbaac093a8fca1e18f8bc37f69e72f8396ac1f770f7aec8b9.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p_meth_pearson_graph_pruned_knn</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">print_graph_info</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned_knn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 381
Number of edges: 3135
Sample nodes: [&#39;TCGA-53-7813&#39;, &#39;TCGA-44-A479&#39;, &#39;TCGA-55-7911&#39;, &#39;TCGA-49-6761&#39;, &#39;TCGA-97-7547&#39;, &#39;TCGA-50-5044&#39;, &#39;TCGA-64-5815&#39;, &#39;TCGA-64-5774&#39;, &#39;TCGA-97-A4LX&#39;, &#39;TCGA-44-A47B&#39;]
Sample edges: [(&#39;TCGA-53-7813&#39;, &#39;TCGA-80-5608&#39;, {&#39;weight&#39;: 0.8771039200330979}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-55-8615&#39;, {&#39;weight&#39;: 0.8753812806142516}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-78-7159&#39;, {&#39;weight&#39;: 0.8660791358358113}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-05-4390&#39;, {&#39;weight&#39;: 0.862051412239931}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-62-A46S&#39;, {&#39;weight&#39;: 0.8619682737544224}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-69-7973&#39;, {&#39;weight&#39;: 0.861408811484938}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-95-A4VP&#39;, {&#39;weight&#39;: 0.8606918998935026}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-50-7109&#39;, {&#39;weight&#39;: 0.8544255697033839}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-44-A479&#39;, {&#39;weight&#39;: 0.8521551540449519}), (&#39;TCGA-53-7813&#39;, &#39;TCGA-MP-A4TD&#39;, {&#39;weight&#39;: 0.8474804112652542})]
No self-loops in the graph.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/c537ea57170a4616622b7af01daa8081eed4e9a38e3957caf8df60c39c42d670.png" src="../_images/c537ea57170a4616622b7af01daa8081eed4e9a38e3957caf8df60c39c42d670.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;patient_meth_network_knn.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="setting-up-pip-network-using-string">
<h1>Setting up PIP network using String<a class="headerlink" href="#setting-up-pip-network-using-string" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a concatenated list of gene names as strings</span>
<span class="c1"># note we are taking gene names from the &#39;gene&#39; column of the dataframe we obtained from the EWAS catalog and filtered</span>
<span class="c1"># to common CpG and Gene sites above. The join function then concatenates them using the &#39;%0D&#39; string</span>
<span class="c1"># to stitch them all together. This string will be used to help us build the API query URL.</span>
<span class="n">external_gene_names</span> <span class="o">=</span> <span class="s1">&#39;%0D&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">df_filt</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">300</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()))</span>

<span class="c1"># pass the list of Gene Names to the String-DB API return the String-IDs</span>
<span class="c1"># we first form the query url using the &#39;get_string_ids&#39; API function which takes a list of identifiers and</span>
<span class="c1"># converts them into the internal String-DB accession IDs. This massively speeds up the search and allows us to</span>
<span class="c1"># search for more than 10 at once which is an API restriction for other API functions if String-DB internal accessions </span>
<span class="c1"># aren&#39;t used.</span>
<span class="n">query_url</span> <span class="o">=</span> <span class="s1">&#39;https://string-db.org/api/tsv-no-header/get_string_ids?identifiers=&#39;</span><span class="o">+</span><span class="n">external_gene_names</span><span class="o">+</span><span class="s1">&#39;&amp;species=9606&amp;format=only-ids&#39;</span>

<span class="c1"># use the urllib library to retrieve the String-DB internal IDs</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">ul</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="n">query_url</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

<span class="c1"># now we want to query String-DB to retrieve interactions from this list of String-DB IDs</span>
<span class="c1"># we create a concatenated list of stringdbIDs in much the same way as above for the Entrez Gene IDs</span>
<span class="n">stringdbIDs</span> <span class="o">=</span> <span class="s1">&#39;%0D&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>

<span class="c1"># again we build the query for interactions using the String-DB IDs</span>
<span class="n">query_url</span> <span class="o">=</span> <span class="s1">&#39;https://string-db.org/api/tsv/network?identifiers=&#39;</span><span class="o">+</span><span class="n">stringdbIDs</span><span class="o">+</span><span class="s1">&#39;&amp;species=9606&#39;</span>

<span class="c1"># again using urllib to retrieve the interactions these are returned in a standard tab delimied text format</span>
<span class="n">interactions</span> <span class="o">=</span> <span class="n">ul</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="n">query_url</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>

<span class="c1"># we need to split the result by these &#39;tabs&#39; (\t - is used to identfy them)</span>
<span class="n">int_test</span> <span class="o">=</span> <span class="p">[</span><span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">interaction</span> <span class="ow">in</span> <span class="n">interactions</span><span class="p">]</span>

<span class="c1"># we extract the field names from the first row</span>
<span class="n">column_names</span> <span class="o">=</span> <span class="n">int_test</span><span class="p">[:</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># create a Pandas dataframe of the interaction data we have just retrieved from String-DB</span>
<span class="n">interactions_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">int_test</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">column_names</span><span class="p">)</span>

<span class="c1"># delete the first row that held the fieldnames but we no longer need</span>
<span class="n">interactions_df</span> <span class="o">=</span> <span class="n">interactions_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># remove any duplicate rows</span>
<span class="n">final_interactions</span> <span class="o">=</span> <span class="n">interactions_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

<span class="c1"># show the top of the protein-protein interaction table</span>
<span class="n">final_interactions</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">save_graphs</span><span class="p">(</span><span class="n">correlation_matrices_dict</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">,</span> <span class="n">matrix_names</span><span class="p">):</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">data_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">data_dir</span><span class="p">)</span>
    
    <span class="c1"># Define parameter combinations</span>
    <span class="n">thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]</span>
    <span class="n">k_values</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">10</span><span class="p">]</span>
    <span class="n">degree_thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">keep_largest_component_options</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">matrix_name</span> <span class="ow">in</span> <span class="n">matrix_names</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">matrix_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">correlation_matrices_dict</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matrix </span><span class="si">{</span><span class="n">matrix_name</span><span class="si">}</span><span class="s2"> not found in the provided dictionary. Skipping...&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
        
        <span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">correlation_matrices_dict</span><span class="p">[</span><span class="n">matrix_name</span><span class="p">]</span>
        
        <span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
            <span class="c1"># Create the initial graph from the correlation matrix</span>
            <span class="n">initial_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="n">threshold</span><span class="p">)</span>
            
            <span class="k">for</span> <span class="n">degree_threshold</span> <span class="ow">in</span> <span class="n">degree_thresholds</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">keep_largest_component</span> <span class="ow">in</span> <span class="n">keep_largest_component_options</span><span class="p">:</span>
                    <span class="c1"># Clean the graph</span>
                    <span class="n">cleaned_graph</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">initial_graph</span><span class="p">,</span> <span class="n">degree_threshold</span><span class="o">=</span><span class="n">degree_threshold</span><span class="p">,</span> <span class="n">keep_largest_component</span><span class="o">=</span><span class="n">keep_largest_component</span><span class="p">)</span>
                    
                    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">k_values</span><span class="p">:</span>
                        <span class="c1"># Sparsify the graph using KNN</span>
                        <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">cleaned_graph</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
                        
                        <span class="c1"># Calculate the average clustering coefficient</span>
                        <span class="n">avg_clustering</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">average_clustering</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="p">)</span>
                        
                        <span class="c1"># Construct the filename with relevant information</span>
                        <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">matrix_name</span><span class="si">}</span><span class="s2">_network_threshold_</span><span class="si">{</span><span class="n">threshold</span><span class="si">}</span><span class="s2">_k_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">_degree_</span><span class="si">{</span><span class="n">degree_threshold</span><span class="si">}</span><span class="s2">_largest_</span><span class="si">{</span><span class="n">keep_largest_component</span><span class="si">}</span><span class="s2">_clustering_</span><span class="si">{</span><span class="n">avg_clustering</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">.gml&quot;</span>
                        <span class="n">filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                        
                        <span class="c1"># Save the graph to a GML file</span>
                        <span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="p">,</span> <span class="n">filepath</span><span class="p">)</span>
                        
                        <span class="c1"># Print the details</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved: </span><span class="si">{</span><span class="n">filepath</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">correlation_matrices_dict</span><span class="p">,</span> <span class="n">matrix_names</span><span class="p">):</span>
    <span class="n">data_dir</span> <span class="o">=</span> <span class="s1">&#39;../../ismb_data/05072024/GE&#39;</span>  <span class="c1"># Adjust the directory path as needed</span>
    <span class="n">save_graphs</span><span class="p">(</span><span class="n">correlation_matrices_dict</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">,</span> <span class="n">matrix_names</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="c1"># correlation_matrices_dict = {</span>
    <span class="c1">#     &#39;pearson&#39;: pd.DataFrame(),  # Replace with actual data</span>
    <span class="c1">#     &#39;spearman&#39;: pd.DataFrame()  # Replace with actual data if needed</span>
    <span class="c1"># }</span>
    <span class="n">matrix_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span>  <span class="c1"># Replace with the desired correlation matrix names</span>
    <span class="n">main</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">,</span> <span class="n">matrix_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./Section1"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../Section1_intro.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Section 1 - Network Basics</p>
      </div>
    </a>
    <a class="right-next"
       href="Part2.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Part 2 - Patient similarity networks</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Part 1 - Introduction to Biomedical Networks</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structuring-biomedical-data-using-networks-workshop-network-generation">Structuring Biomedical Data Using Networks  workshop - Network Generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#biomedical-networks">1. Biomedical Networks</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#libraries">1.1 Libraries</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcga-luad-project">1.2 TCGA-LUAD Project</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata-analysis">1.3 Metadata Analysis</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-data-analysis">Expression data analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-filtering-and-preprocessing">2. Data Filtering and Preprocessing</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysing-gene-expression-levels">2.1 Analysing Gene Expression Levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholding-based-on-gene-expression-variance">2.3 Thresholding Based on Gene Expression Variance</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#task-1-variance-based-gene-filtering">Task 1: Variance-Based Gene Filtering</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-calculate-variance">Step 1: Calculate Variance</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-set-a-variance-threshold">Step 2: Set a Variance Threshold</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-filter-genes">Step 3: Filter Genes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-retention-at-different-variance-levels">2.4 Gene Retention at Different Variance Levels</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-symbol-conversion-ensg-to-hgcn">3. Gene Symbol Conversion: ENSG to HGCN</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-data-investigation">Further Data Investigation</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#correlation-matrix-calculation">4. Correlation Matrix Calculation</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-construction-from-correlation-matrices">5. Graph Construction from Correlation Matrices</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#network-cleaning">6. Network Cleaning</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#sparsification-methods">7. Sparsification methods</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#describing-highly-connected-nodes">8. Describing Highly Connected Nodes</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#interactive-visualization">Interactive Visualization</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-add-hover-info-to-plotly-plot-gene-info">TODO: add hover info to plotly plot (gene info)</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-introduce-clustering-kmeans-etc">TODO: introduce clustering kmeans. etc</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#storing-generated-networks">Storing generated networks</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#patient-similarity-network">Patient similarity network</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#dnameth-pn">DNAmeth PN</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-pip-network-using-string">Setting up PIP network using String</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By https://biomedicalinformaticsgroup.github.io/
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>