

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Part 1: Over-representation and Enrichment Analysis &#8212; ISMB Tutorial Book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Section2/Part1';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Part 2: Similarity Network Fusion" href="Part2.html" />
    <link rel="prev" title="Section 2: Network Analysis" href="../Section2_intro.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="ISMB Tutorial Book - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="ISMB Tutorial Book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    An Introduction to Multi Modal Analysis using Networks
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../GettingStarted.html">Getting Started</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section1_intro.html">Section 1 - Network Basics</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section1/Part1.html">Part 1 - Introduction to Biomedical Networks</a></li>
















<li class="toctree-l2"><a class="reference internal" href="../Section1/Part2.html">Part 2 - Patient similarity networks</a></li>

</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../Section2_intro.html">Section 2: Network Analysis</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Part 1: Over-representation and Enrichment Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="Part2.html">Part 2: Similarity Network Fusion</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section3_intro.html">Section 3 - SNF + GNN’s</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part1.html"><font color="darkblue"> Part 1 : Message Passing </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part2.html"><font color="darkblue"> Part 2 : Problem Formulation / Patient Similarity Network Generation </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part3.html"><font color="darkblue"> Part 3 : Smoking Network Classification using Graph Convolutional Networks</font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part4.html">Part 4 : Similarity Network Fusion with Graph Convolutional Networks</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ConceptGlossary.html">Concept Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions_definitions.html">Custom Function Definitions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../feedback_survey.html">Feedback Survey</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/biomedicalinformaticsgroup/networks_ismb24.git" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/biomedicalinformaticsgroup/networks_ismb24.git/issues/new?title=Issue%20on%20page%20%2FSection2/Part1.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Section2/Part1.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Part 1: Over-representation and Enrichment Analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-networks-from-section-1">0. Load networks from Section 1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-function-draw-network-with-node-attrs-draws-a-network-with-nodes-colored-and-or-shaped-based-on-their-attributes-if-communities-are-provided-nodes-are-colored-by-their-community-memberships-a-legend-is-added-to-indicate-the-mapping-of-attributes-to-colors-and-shapes">Custom function <code class="docutils literal notranslate"><span class="pre">draw_network_with_node_attrs()</span></code> draws a network with nodes colored and/or shaped based on their attributes. If communities are provided, nodes are colored by their community memberships. A legend is added to indicate the mapping of attributes to colors and shapes.</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#we-will-use-cancer-gene-calatogue-cgc-from-the-catalogue-of-somatic-mutations-in-cancer-cosmic-to-annotate-genes-in-our-network">We will use cancer gene calatogue (CGC) from the Catalogue Of Somatic Mutations In Cancer (COSMIC) to annotate genes in our network.</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-network-analysis-with-g-exp">1. Gene Network Analysis with <code class="docutils literal notranslate"><span class="pre">G_exp</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#over-representation-analysis-ora">1.1 Over Representation Analysis (ORA)</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-correlation-network">1.1.1 ORA on gene correlation network</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-clusters">1.1.1 ORA on gene clusters</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-the-communities">Visualising the Communities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-ora-on-each-community">Perform ORA on Each Community</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-set-enrichment-analysis-gsea">1.2 Gene Set Enrichment Analysis (GSEA)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status">Assign classes based on phenotypic attributes in metadata (e.g., smoking status)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-gsea-results">Visualising GSEA Results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#barcode-enrichment-plot">Barcode Enrichment Plot</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heatmap-visualisation">Heatmap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clustermap-visualisation">Clustermap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dot-plot-visualisation">Dot Plot Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gsea-on-clusters">GSEA on Clusters</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="part-1-over-representation-and-enrichment-analysis">
<h1>Part 1: Over-representation and Enrichment Analysis<a class="headerlink" href="#part-1-over-representation-and-enrichment-analysis" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">urllib</span>
<span class="kn">import</span> <span class="nn">urllib</span> <span class="k">as</span> <span class="nn">ul</span>
<span class="kn">import</span> <span class="nn">urllib.request</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">import</span> <span class="nn">ast</span>
<span class="kn">from</span> <span class="nn">prettytable</span> <span class="kn">import</span> <span class="n">PrettyTable</span>
<span class="kn">import</span> <span class="nn">gseapy</span> <span class="k">as</span> <span class="nn">gp</span>
<span class="kn">from</span> <span class="nn">palettable</span> <span class="kn">import</span> <span class="n">wesanderson</span>

<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_colwidth&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span>
<span class="n">current_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;__file__&#39;</span><span class="p">))</span>
<span class="n">parent_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">current_dir</span><span class="p">,</span> <span class="s1">&#39;..&#39;</span><span class="p">))</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parent_dir</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parent_dir</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="load-networks-from-section-1">
<h2>0. Load networks from Section 1<a class="headerlink" href="#load-networks-from-section-1" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Gene correlation network</p></li>
<li><p>Patient network from TCGA gene expression data</p></li>
<li><p>Patient network from TCGA DNA methylation data</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">intermediate_data_dir</span> <span class="o">=</span> <span class="s1">&#39;/data/intermediate/&#39;</span>
<span class="n">raw_data_dir</span> <span class="o">=</span> <span class="s1">&#39;/data/raw/&#39;</span>

<span class="c1"># Define paths to .gml network files from Section 1</span>
<span class="c1"># These paths point to the Gene correlation network, Patient network from (1) TCGA gene expression data and (2) TCGA DNA methylation data.</span>
<span class="n">G_gxp_path</span> <span class="o">=</span> <span class="n">intermediate_data_dir</span> <span class="o">+</span> <span class="s1">&#39;section2_networks_v1/gene_coexpression_network.gml&#39;</span>

<span class="c1"># Load the GML graphs into NetworkX graph objects</span>
<span class="c1"># nx.read_gml() function reads a graph from a GML file</span>
<span class="n">G_gxp</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">read_gml</span><span class="p">(</span><span class="n">G_gxp_path</span><span class="p">)</span>  <span class="c1"># Gene correlation network</span>

<span class="c1"># Get all nodes in each graph</span>
<span class="c1"># The nodes represent genes or patients depending on the network</span>
<span class="n">G_gxp_nodes_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">G_gxp</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>  <span class="c1"># Nodes in the gene correlation network</span>

<span class="c1"># Define paths to the raw TCGA datasets</span>
<span class="c1"># tcga_dnam_path = &#39;section2_data/ISMB_TCGA_DNAm.pkl&#39;  # TCGA DNA methylation data</span>
<span class="n">tcga_gxp_path</span> <span class="o">=</span> <span class="n">raw_data_dir</span> <span class="o">+</span> <span class="s1">&#39;ISMB_TCGA_GE.pkl&#39;</span>  <span class="c1"># TCGA Gene expression data</span>

<span class="c1"># Load the gene expression dataset</span>
<span class="c1"># pd.read_pickle() function loads a pickled pandas DataFrame or Series</span>
<span class="n">tcga_gxp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="n">tcga_gxp_path</span><span class="p">)</span>

<span class="c1"># For this example, we&#39;ll use a CSV file that includes gene symbols</span>
<span class="c1"># pd.read_csv() function loads a CSV file into a pandas DataFrame</span>
<span class="n">tcga_gxp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">intermediate_data_dir</span> <span class="o">+</span> <span class="s1">&#39;tcga_ge_df_symbols_t.csv&#39;</span><span class="p">)</span> <span class="c1"># Dataset with gene symbols</span>
<span class="c1"># Set &#39;GENES&#39; column as the index for easy access to gene-specific data</span>
<span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;GENES&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Extract metadata from the gene expression dataset</span>
<span class="c1"># Metadata might include information such as patient IDs, sample conditions, etc.</span>
<span class="n">tcga_gxp_meta</span> <span class="o">=</span> <span class="n">tcga_gxp</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">]</span>

<span class="c1"># Print the number of nodes in each network</span>
<span class="c1"># This provides a quick overview of the size of each network</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of nodes in gene correlation network: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">G_gxp_nodes_list</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="custom-function-draw-network-with-node-attrs-draws-a-network-with-nodes-colored-and-or-shaped-based-on-their-attributes-if-communities-are-provided-nodes-are-colored-by-their-community-memberships-a-legend-is-added-to-indicate-the-mapping-of-attributes-to-colors-and-shapes">
<h3>Custom function <code class="docutils literal notranslate"><span class="pre">draw_network_with_node_attrs()</span></code> draws a network with nodes colored and/or shaped based on their attributes. If communities are provided, nodes are colored by their community memberships. A legend is added to indicate the mapping of attributes to colors and shapes.<a class="headerlink" href="#custom-function-draw-network-with-node-attrs-draws-a-network-with-nodes-colored-and-or-shaped-based-on-their-attributes-if-communities-are-provided-nodes-are-colored-by-their-community-memberships-a-legend-is-added-to-indicate-the-mapping-of-attributes-to-colors-and-shapes" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">functions</span> <span class="kn">import</span> <span class="n">draw_network_with_node_attrs</span>

<span class="c1"># Args:</span>
<span class="c1">#     G (networkx.Graph): The graph to be drawn.</span>
<span class="c1">#     node_attributes (dict): A dictionary where keys are node names and values are dictionaries of attributes.</span>
<span class="c1">#     communities (List[List[Any]], optional): A list where each sublist contains the nodes belonging to a community. Default is None.</span>
<span class="c1">#     title (str, optional): The title of the plot. Default is &#39;Network Visualization&#39;.</span>
<span class="c1">#     color_attr (str, optional): Node attribute to color nodes by. Default is None.</span>
<span class="c1">#     shape_attr (str, optional): Node attribute to shape nodes by. Default is None.</span>
<span class="c1">#     figsize (tuple, optional): The size of the figure. Default is (20, 10).</span>
<span class="c1">#     layout (str, optional): The layout algorithm for positioning nodes (&#39;spring&#39;, &#39;circular&#39;, etc.). Default is &#39;spring&#39;.</span>
<span class="c1">#     cmap_name (str, optional): The name of the colormap to use for coloring. Default is &#39;tab20&#39;.</span>
<span class="c1">#     with_labels (bool, optional): Whether to draw labels for the nodes. Default is False.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the shape of the dataframe</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Shape of the dataframe &#39;tcga_gxp_df&#39; (rows, columns):&quot;</span><span class="p">)</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="nb">print</span><span class="p">()</span>

<span class="c1"># List the columns in the dataframe</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">List of columns in the dataframe &#39;tcga_gxp_df&#39;:&quot;</span><span class="p">)</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="nb">print</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="we-will-use-cancer-gene-calatogue-cgc-from-the-catalogue-of-somatic-mutations-in-cancer-cosmic-to-annotate-genes-in-our-network">
<h3>We will use <a class="reference external" href="https://cancer.sanger.ac.uk/census">cancer gene calatogue (CGC)</a> from the Catalogue Of Somatic Mutations In Cancer (COSMIC) to annotate genes in our network.<a class="headerlink" href="#we-will-use-cancer-gene-calatogue-cgc-from-the-catalogue-of-somatic-mutations-in-cancer-cosmic-to-annotate-genes-in-our-network" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">'Tier'</span></code></p>
<ul>
<li><p>To be classified into Tier 1, a gene must possess a documented activity relevant to cancer, along with evidence of mutations in cancer which change the activity of the gene product in a way that promotes oncogenic transformation.</p></li>
<li><p>Tier 2 consists of genes with strong indications of a role in cancer but with less extensive available evidence.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">'Hallmark'</span></code></p>
<ul>
<li><p>New overviews of cancer gene function focused on hallmarks of cancer pull together manually curated information on the function of proteins coded by cancer genes and summarise the data in simple graphical form. They present a condensed overview of most relevant facts with quick access to the literature source, and define whether a gene has a stimulating or suppressive effect via individual cancer hallmarks.</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cancer_genes_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/Census_allFri Jul  5 14_32_40 2024.csv&#39;</span>
<span class="n">cancer_genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cancer_genes_path</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cancer_genes_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="n">cancer_genes_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cancer_genes_df</span> <span class="o">=</span> <span class="n">cancer_genes_df</span><span class="p">[</span><span class="n">cancer_genes_df</span><span class="p">[</span><span class="s1">&#39;Tier&#39;</span><span class="p">]</span><span class="o">==</span><span class="mi">1</span><span class="p">]</span>
<span class="n">cancer_genes</span> <span class="o">=</span> <span class="n">cancer_genes_df</span><span class="p">[</span><span class="s1">&#39;Gene Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># Create the dictionary with node labels as keys and boolean as values</span>
<span class="c1"># Complete the dictionary comprehension to check if each node is in the list of cancer genes</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">found_in_cancer_genes</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="p">:</span> <span class="nb">bool</span><span class="p">(</span><span class="n">node</span> <span class="ow">in</span> <span class="n">cancer_genes</span><span class="p">)</span> <span class="p">}</span>

<span class="n">node_attributes</span><span class="o">=</span><span class="p">{}</span>
<span class="n">node_attributes</span><span class="p">[</span><span class="s1">&#39;cancer_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">found_in_cancer_genes</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">node_attributes</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="gene-network-analysis-with-g-exp">
<h2>1. Gene Network Analysis with <code class="docutils literal notranslate"><span class="pre">G_exp</span></code><a class="headerlink" href="#gene-network-analysis-with-g-exp" title="Permalink to this heading">#</a></h2>
<p>We are going to use <code class="docutils literal notranslate"><span class="pre">'KEGG_2021_Human'</span></code> as the gene set. KEGG (Kyoto Encyclopedia of Genes and Genomes) is a knowledge base for systematic analysis of gene functions, linking genomic information with higher order functional information <a class="reference external" href="https://doi.org/10.1093/nar/28.1.27">(Kanehisa et al., 2000)</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Alternative gene sets like &#39;MSigDB_Hallmark_2020&#39; can also be used.</span>
<span class="n">gene_sets</span> <span class="o">=</span> <span class="s1">&#39;KEGG_2021_Human&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can also retrieve and display the list of available gene sets</span>
<span class="n">gene_set_list</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">get_library_name</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_set_list</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="over-representation-analysis-ora">
<h3>1.1 Over Representation Analysis (ORA)<a class="headerlink" href="#over-representation-analysis-ora" title="Permalink to this heading">#</a></h3>
<p>Over-representation analysis (ORA) is a method used to identify which predefined gene sets are disproportionately represented in a given set of genes compared to what would be expected by random chance <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2615629/">(Huang et al., 2009)</a>. We recommend using Over-representation analysis (ORA) only when Gene Set Enrichment Analysis (GSEA) is not suitable. Although we are using the <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btac757"><code class="docutils literal notranslate"><span class="pre">gseapy</span></code></a> library for ORA in this the tutorial, it’s important to note that ORA and GSEA are distinct methods.</p>
<section id="ora-on-gene-correlation-network">
<h4>1.1.1 ORA on gene correlation network<a class="headerlink" href="#ora-on-gene-correlation-network" title="Permalink to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform ORA on all nodes in the gene correlation network</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">enr_all_nodes</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the top 10 enrichment results</span>

<span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="ora-on-gene-clusters">
<h4>1.1.1 ORA on gene clusters<a class="headerlink" href="#ora-on-gene-clusters" title="Permalink to this heading">#</a></h4>
<p>Clustering on Gene Correlation Network</p>
<ul class="simple">
<li><p>We use community detection algorithm to identify communities in the network:</p>
<ul>
<li><p>The <code class="docutils literal notranslate"><span class="pre">greedy_modularity_communities()</span></code> function in <code class="docutils literal notranslate"><span class="pre">networkx</span></code> implements a community detection algorithm that optimises modularity using a greedy approach. It iteratively merges pairs of nodes or communities that result in the largest increase in modularity until no further improvement is possible. Modularity measures the density of links inside communities compared to links between communities, aiming to maximise this value to identify densely connected groups within the network.</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster the gene correlation network using the greedy modularity communities algorithm</span>
<span class="n">communities</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">algorithms</span><span class="o">.</span><span class="n">community</span><span class="o">.</span><span class="n">modularity_max</span><span class="o">.</span><span class="n">greedy_modularity_communities</span><span class="p">(</span><span class="n">G_gxp</span><span class="p">)</span>

<span class="c1"># Print the number of detected communities</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The network has </span><span class="si">{}</span><span class="s1"> communities.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create subgraphs for each community</span>
<span class="n">subgraphs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">community</span> <span class="ow">in</span> <span class="n">communities</span><span class="p">:</span>
    
    <span class="n">subgraphs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">G_gxp</span><span class="o">.</span><span class="n">subgraph</span><span class="p">(</span><span class="n">community</span><span class="p">))</span>

<span class="c1"># Print the number of nodes in each subgraph</span>
<span class="c1"># And draw the subgraph</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">subgraph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">subgraphs</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Community </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1"> has </span><span class="si">{</span><span class="n">subgraph</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">()</span><span class="si">}</span><span class="s1"> nodes.&#39;</span><span class="p">)</span>
    
    <span class="c1">### YOUR CODE HERE ###</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> <span class="c1"># Forces the plot to be displayed</span>
    
    
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="visualising-the-communities">
<h3>Visualising the Communities<a class="headerlink" href="#visualising-the-communities" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualise the gene correlation network with community memberships</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">draw_network_with_node_attrs</span><span class="p">()</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="perform-ora-on-each-community">
<h3>Perform ORA on Each Community<a class="headerlink" href="#perform-ora-on-each-community" title="Permalink to this heading">#</a></h3>
<p>Performing ORA on individual clusters can help in understanding distinct biological significance of each cluster, revealing how certain pathways or functions are associated with specific subsets of genes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to perform ORA on a given list of genes</span>
<span class="k">def</span> <span class="nf">communityORA</span><span class="p">(</span><span class="n">genes</span><span class="p">):</span>
    <span class="n">enr</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">gene_sets</span><span class="o">=</span><span class="p">[</span><span class="n">gene_sets</span><span class="p">],</span> <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">enr</span>

<span class="c1"># 1. Convert communities to lists for ORA 2. Sort the list of communities by their length in descending order</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">communities</span> <span class="o">=</span> 
<span class="n">communities</span> <span class="o">=</span> 

<span class="c1"># Perform ORA for three selected communities</span>
<span class="n">community1_enr</span> <span class="o">=</span> <span class="n">communityORA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">community2_enr</span> <span class="o">=</span> <span class="n">communityORA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">community3_enr</span> <span class="o">=</span> <span class="n">communityORA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

<span class="c1"># Print the top 10 results for each community</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">PrettyTable</span><span class="p">()</span>
<span class="n">x</span><span class="o">.</span><span class="n">field_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Community 1&quot;</span><span class="p">,</span> <span class="s2">&quot;Community 2&quot;</span><span class="p">,</span> <span class="s2">&quot;Community 3&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="n">x</span><span class="o">.</span><span class="n">add_row</span><span class="p">([</span><span class="n">community1_enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">community2_enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">community3_enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="gene-set-enrichment-analysis-gsea">
<h3>1.2 Gene Set Enrichment Analysis (GSEA)<a class="headerlink" href="#gene-set-enrichment-analysis-gsea" title="Permalink to this heading">#</a></h3>
<p>Gene Set Enrichment Analysis (GSEA) is a genome-wide expression analysis method designed to interpret expression profiles focusing on pre-defined gene sets <a class="reference external" href="https://doi.org/10.1073/pnas.0506580102">(Subramanian et al., 2005)</a>. These gene sets are curated based on prior biological knowledge, such as published information about biochemical pathways or patterns of coexpression observed in previous experimental studies. The genes can be ordered in a ranked list, according to their differential expression between the classes. The primary objective of GSEA is to assess whether the genes within a given gene set tend to occur toward the top (or bottom) of the ranked list. This ranking is based on the correlation between gene expression and a particular phenotypic class distinction. By evaluating the distribution of gene set members within the ranked list, GSEA identifies whether the set is correlated with the phenotypic class, thus providing insights into underlying biological mechanisms. This method contrasts with traditional single-gene analysis by focusing on the collective behavior of gene sets, thereby uncovering biologically significant patterns that might be overlooked when examining individual genes in isolation. We use <code class="docutils literal notranslate"><span class="pre">gseapy</span></code> to perform GSEA with the gene set <code class="docutils literal notranslate"><span class="pre">KEGG_2021_Human</span></code>.</p>
<!-- (https://pnnl-comp-mass-spec.github.io/proteomics-data-analysis-tutorial/gsea.html) for referencing.  -->
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the first five rows of the dataframe &quot;tcga_gxp_df&quot;</span>
<span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcga_gxp_meta</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status">
<h3>Assign classes based on phenotypic attributes in metadata (e.g., smoking status)<a class="headerlink" href="#assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dictionary to store the class assignments for each sample</span>
<span class="n">classes</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Iterate through the columns (samples)</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">G_gxp_nodes_list</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span> <span class="c1"># &quot;.loc[G_gxp_nodes_list]&quot; grabs rows of specified gene symbols</span>
    
    <span class="c1"># Assign class labels based on the smoking status from the metadata</span>
    
    <span class="k">if</span> <span class="n">tcga_gxp_meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;Smoked&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Smoker&#39;</span><span class="p">:</span>
        <span class="c1">### YOUR CODE HERE ###</span>
    <span class="k">elif</span> <span class="n">tcga_gxp_meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;Smoked&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Never&#39;</span><span class="p">:</span>
        <span class="c1">### YOUR CODE HERE ###</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">classes</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">G_gxp_nodes_list</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform GSEA using the prepared data and class assignments</span>
<span class="n">gs_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">gsea</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">G_gxp_nodes_list</span><span class="p">],</span> <span class="n">gene_sets</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">,</span> <span class="bp">cls</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span> <span class="n">permutation_num</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;signal_to_noise&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>

<span class="c1"># Display the top results from the GSEA</span>

<span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualising-gsea-results">
<h3>Visualising GSEA Results<a class="headerlink" href="#visualising-gsea-results" title="Permalink to this heading">#</a></h3>
<p>Once you have performed GSEA, the next step is to visualise the results. Visualisation helps in interpreting the biological roles of the enriched gene sets. Here, we visualise GSEA results with Barcode Enrichment Plot, Heatmap, Clustermap, and Dot Plot.</p>
</section>
</section>
<section id="barcode-enrichment-plot">
<h2>Barcode Enrichment Plot<a class="headerlink" href="#barcode-enrichment-plot" title="Permalink to this heading">#</a></h2>
<p>Barcode Enrichment Plot shows the positions of members of a given gene set in a ranked list of enrichment scores for the top enriched terms. The scores are ranked left to right from smallest to largest. The ranked scores are represented by a shaded bar, forming a pattern like a barcode.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract the enriched terms from the GSEA results. The terms represent pathways or functional categories that are significantly enriched in the dataset.</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">terms</span> <span class="o">=</span> 

<span class="c1"># Plot the top 5 enriched terms</span>
<span class="c1"># The plot function visualizes the enrichment results</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">terms</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">show_ranking</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">legend_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;loc&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">0</span><span class="p">)})</span>
</pre></div>
</div>
</div>
</div>
<p>You can view and extract leading-edge genes from GSEA results. Leading-edge genes are the subset of genes that contribute most to the enrichment score.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># View leading-edge genes from the GSEA results. </span>
<span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[[</span><span class="s1">&#39;Term&#39;</span><span class="p">,</span> <span class="s1">&#39;Lead_genes&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="heatmap-visualisation">
<h2>Heatmap Visualisation<a class="headerlink" href="#heatmap-visualisation" title="Permalink to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">gseapy</span></code> provides a heatmap function to visualise the expression levels of the leading-edge genes. The heatmap provides a visual representation of how these genes are expressed across different samples in relation to their assigned phenotypic classes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the heatmap function from gseapy</span>
<span class="kn">from</span> <span class="nn">gseapy</span> <span class="kn">import</span> <span class="n">heatmap</span>

<span class="c1"># Select the index of the term to visualize</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>

<span class="c1"># Extract the genes contributing to the enrichment of the selected term</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">genes</span> <span class="o">=</span> 
<span class="nb">print</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate a heatmap of the expression levels of the leading-edge genes</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">heatmap</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">gs_res</span><span class="o">.</span><span class="n">heatmat</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">],</span> <span class="n">z_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">terms</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

<span class="c1"># Update the x-tick labels with the class labels</span>
<span class="n">xtick_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">classes</span><span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">get_text</span><span class="p">()]</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()]</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xtick_labels</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="clustermap-visualisation">
<h2>Clustermap Visualisation<a class="headerlink" href="#clustermap-visualisation" title="Permalink to this heading">#</a></h2>
<p>The function <code class="docutils literal notranslate"><span class="pre">clustermap</span></code> from <code class="docutils literal notranslate"><span class="pre">seaborn</span></code> is used to create a clustered heatmap. It not only shows the expression levels of the leading-edge genes but also clusters them based on similarity, providing additional insights into gene expression patterns. The cluster map includes dendrograms, which show the hierarchical clustering of both genes and samples, helping to identify groups of co-expressed genes and similar samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the clustermap function from seaborn</span>
<span class="kn">from</span> <span class="nn">seaborn</span> <span class="kn">import</span> <span class="n">clustermap</span>

<span class="c1"># Select the index of the term to visualize</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># Extract the genes contributing to the enrichment of the selected term</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Lead_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>

<span class="c1"># Extract the relevant subset of the heatmap data</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">data</span> <span class="o">=</span> 

<span class="c1"># Rename the columns based on the class assignments</span>
<span class="n">data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Generate a cluster map of the expression levels of the leading-edge genes</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">clustermap</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>  <span class="c1"># The data to cluster</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">,</span>  <span class="c1"># Clustering method</span>
    <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span>  <span class="c1"># Distance metric</span>
    <span class="n">z_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>  <span class="c1"># Standardize the data along the rows</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>  <span class="c1"># Size of the figure</span>
    <span class="n">dendrogram_ratio</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>  <span class="c1"># Ratio of the dendrogram</span>
    <span class="n">colors_ratio</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span>  <span class="c1"># Ratio of the colors</span>
    <span class="n">cbar_pos</span><span class="o">=</span><span class="p">(</span><span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># Position of the color bar</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="dot-plot-visualisation">
<h2>Dot Plot Visualisation<a class="headerlink" href="#dot-plot-visualisation" title="Permalink to this heading">#</a></h2>
<p>Use the <code class="docutils literal notranslate"><span class="pre">dotplot</span></code> function in <code class="docutils literal notranslate"><span class="pre">gseapy</span></code> to create a visual representation of the GSEA results. Here we use “FDR q-val” to determine the dot sizes, which represents the false discovery rate adjusted p-values. We display normalised enrichment score (NES) value as the x-axis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the dotplot function from gseapy</span>
<span class="kn">from</span> <span class="nn">gseapy</span> <span class="kn">import</span> <span class="n">dotplot</span>

<span class="c1"># Generate the dot plot for the GSEA results</span>
<span class="c1"># The dotplot function visualizes the enrichment results, focusing on the FDR q-values</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">dotplot</span><span class="p">(</span><span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">,</span>
             <span class="n">column</span><span class="o">=</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">,</span>  <span class="c1"># Column to be used for dot size</span>
             <span class="n">title</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>  <span class="c1"># Title the plot as the chosen gene set name </span>
             <span class="n">cmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">viridis</span><span class="p">,</span>  <span class="c1"># Color map for the dots</span>
             <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>  <span class="c1"># Size of the dots</span>
             <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>  <span class="c1"># Size of the figure</span>
             <span class="n">cutoff</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Cutoff for displaying the terms</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="gsea-on-clusters">
<h2>GSEA on Clusters<a class="headerlink" href="#gsea-on-clusters" title="Permalink to this heading">#</a></h2>
<p>Similarly to ORA, GSEA can also be performed on individual communities after clustering. This allows for a more granular analysis, revealing pathways and functions that are enriched within particular subgroups of the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a function to perform GSEA on a given list of genes (community)</span>
<span class="k">def</span> <span class="nf">communityGSEA</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">gene_sets</span><span class="o">=</span><span class="s1">&#39;KEGG_2021_Human&#39;</span><span class="p">,</span> <span class="n">classes</span><span class="o">=</span><span class="n">classes</span><span class="p">):</span>
    <span class="n">gs_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">gsea</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">],</span> <span class="n">gene_sets</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">,</span> <span class="bp">cls</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span> <span class="n">permutation_num</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;signal_to_noise&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> 
    <span class="k">return</span> <span class="n">gs_res</span>

<span class="c1"># Perform GSEA for three community</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">community1_gsea</span> <span class="o">=</span> <span class="n">communityGSEA</span><span class="p">()</span>
<span class="n">community2_gsea</span> <span class="o">=</span> <span class="n">communityGSEA</span><span class="p">()</span> 
<span class="n">community3_gsea</span> <span class="o">=</span> <span class="n">communityGSEA</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the top results for the first community</span>

<span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract the enriched terms from the GSEA results for the first community</span>
<span class="n">terms</span> <span class="o">=</span> <span class="n">community1_gsea</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Term</span>

<span class="c1"># Plot the top 5 enriched terms for the first community</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">community1_gsea</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">terms</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">show_ranking</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">legend_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;loc&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">0</span><span class="p">)})</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./Section2"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../Section2_intro.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Section 2: Network Analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="Part2.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Part 2: Similarity Network Fusion</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-networks-from-section-1">0. Load networks from Section 1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-function-draw-network-with-node-attrs-draws-a-network-with-nodes-colored-and-or-shaped-based-on-their-attributes-if-communities-are-provided-nodes-are-colored-by-their-community-memberships-a-legend-is-added-to-indicate-the-mapping-of-attributes-to-colors-and-shapes">Custom function <code class="docutils literal notranslate"><span class="pre">draw_network_with_node_attrs()</span></code> draws a network with nodes colored and/or shaped based on their attributes. If communities are provided, nodes are colored by their community memberships. A legend is added to indicate the mapping of attributes to colors and shapes.</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#we-will-use-cancer-gene-calatogue-cgc-from-the-catalogue-of-somatic-mutations-in-cancer-cosmic-to-annotate-genes-in-our-network">We will use cancer gene calatogue (CGC) from the Catalogue Of Somatic Mutations In Cancer (COSMIC) to annotate genes in our network.</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-network-analysis-with-g-exp">1. Gene Network Analysis with <code class="docutils literal notranslate"><span class="pre">G_exp</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#over-representation-analysis-ora">1.1 Over Representation Analysis (ORA)</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-correlation-network">1.1.1 ORA on gene correlation network</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-clusters">1.1.1 ORA on gene clusters</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-the-communities">Visualising the Communities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-ora-on-each-community">Perform ORA on Each Community</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-set-enrichment-analysis-gsea">1.2 Gene Set Enrichment Analysis (GSEA)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status">Assign classes based on phenotypic attributes in metadata (e.g., smoking status)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-gsea-results">Visualising GSEA Results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#barcode-enrichment-plot">Barcode Enrichment Plot</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heatmap-visualisation">Heatmap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clustermap-visualisation">Clustermap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dot-plot-visualisation">Dot Plot Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gsea-on-clusters">GSEA on Clusters</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By https://biomedicalinformaticsgroup.github.io/
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>